<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SCMultiomeCNRSR">
<title>Basics of Single-cell Annotation â€¢ SCMultiomeCNRSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Basics of Single-cell Annotation">
<meta property="og:description" content="SCMultiomeCNRSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCMultiomeCNRSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction_to_single-cell.html">Introduction to single-cell pre-processing</a>
    <a class="dropdown-item" href="../articles/basics_of_single-cell_annotation.html">Basics of single-cell annotation</a>
    <a class="dropdown-item" href="../articles/advancedphenotyping.html">Advanced Cell Phenotyping</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Basics of Single-cell Annotation</h1>
                        <h4 data-toc-skip class="author">Carissa Chen<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:cchen@cmri.org.au" class="email"&gt;cchen@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Sharon
Long<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:slong@cmri.org.au" class="email"&gt;slong@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R/blob/HEAD/vignettes/basics_of_single-cell_annotation.Rmd" class="external-link"><code>vignettes/basics_of_single-cell_annotation.Rmd</code></a></small>
      <div class="d-none name"><code>basics_of_single-cell_annotation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>Single-cell annotation is the key step in scRNA-seq analysis to
reveal the heterogeneity present in the data for further downstream
analyses to characterise the specific cell type differences. In this
section of the workshop, we will explore both unsupervised and
supervised techniques for cell type identification, leveraging tools
such as clustering algorithms, marker gene annotation, and machine
learning classifiers. The <em>learning goals</em> of this section is to
understand the different ways cell type annotation can be performed and
under which scenarios it is more preferable to use either approach.</p>
<p><br></p>
<p><span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
<strong>Preparation and assumed knowledge</strong></p>
<ul>
<li><a href="https://doi.org/10.1371/journal.pcbi.1011288" class="external-link">Use of
dimensionality reduction in single-cell genomics</a></li>
<li><a href="https://doi.org/10.15252/msb.20199389" class="external-link">scClassify</a></li>
</ul>
<p><br></p>
<p><span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
<strong>Learning objectives</strong></p>
<ul>
<li>Apply different dimensionality reduction techniques and customise
the visualisation</li>
<li>To perform unsupervised and supervised procedures to annotate cell
types in scRNA-seq data</li>
<li>Think critically about the disadvantages and advantages of each
approach</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="schedule">Schedule<a class="anchor" aria-label="anchor" href="#schedule"></a>
</h3>
<p>Structure of this section:</p>
<table class="table">
<colgroup>
<col width="90%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Data visualisation using dimensionality reduction</td>
<td>Xm</td>
</tr>
<tr class="odd">
<td>Unsupervised clustering and annotation of scRNA-seq data</td>
<td>Xm</td>
</tr>
<tr class="even">
<td>Perform cell classification with <code>scClassify</code>
</td>
<td>Xm</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>We have now preprocessed our single cell data. The next step is to
analyse the data for cell type identification, identification of marker
genes and <em>et cetera</em>.</p>
<p>We will focus on some simple steps of downstream analysis for single
cell RNA-sequencing (scRNA-seq) data as shown below.</p>
<ol style="list-style-type: decimal">
<li>What are the cell types present in the dataset?</li>
<li>What are these cell clusters?</li>
<li>For a gene of interest, how can I visualise the gene expression
distribution?</li>
<li>What are the cell type composition in the data?</li>
</ol>
<p>Two common approaches are used - unsupervised clustering and manual
annotation using marker genes, and annotation by supervised
classification.</p>
<p>Unsupervised clustering is a method that groups cells into clusters
based on their gene expression profiles without prior knowledge of the
cell types. The goal is to identify groupings of cells with similar
transcriptomic profiles which correspond to a single cell type. After
clustering, the resulting groupings are annotated by using known marker
genes that are highly expressed in specific cell types, serving as a
defining signature of a cell type. Alternatively, differential gene
analysis can be performed on the clusters to determine a selection of
top genes that are related to specific cell types to annotate the
cluster.</p>
<p>If there is an appropriate reference single-cell dataset, we can opt
to use a supervised classification algorithm to classify our cells.
Typically, the classification model is trained on a labeled dataset,
where it learns the relationships between cell types and gene expression
profiles. Then, the trained model is applied on a new, unlabeled dataset
to predict the cell type based on prior gene expression information.</p>
<p>In this tutorial we will explore an example of the typical
unsupervised annotation method using <code>Seurat</code>, and then
demonstrate the usage of a supervised classification using
<code>scClassify</code>.</p>
</div>
<div class="section level2">
<h2 id="loading-libraries-and-the-data">Loading libraries and the data<a class="anchor" aria-label="anchor" href="#loading-libraries-and-the-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jrnold.github.io/ggthemes/" class="external-link">ggthemes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/VPetukhov/ggrastr" class="external-link">ggrastr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/" class="external-link">mclust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">presto</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"../data/hippocampus_processed.rda"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">hippocampus_processed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">hippocampus_processed</span><span class="op">)</span></span></code></pre></div>
<p>To simplify our data, we will only take two time points and to
simplify the data, we will only take the continuous cell types which
will be helpful for the next part of the workshop.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">age</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P0"</span>, <span class="st">"P5"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Immature-GC"</span>, <span class="st">"Immature-Pyr"</span>, <span class="st">"Neuroblast"</span>, <span class="st">"nIPC"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">cell_cluster</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">keep</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">convert_to_seurat</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sce</span>, <span class="va">meta</span> <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">drops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">meta</span><span class="op">)</span> </span>
<span>    <span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span>  <span class="kw">else</span> <span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">drops</span>, ,drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>  <span class="va">seur</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">seur</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">convert_to_seurat</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="data-visualisation-using-dimensionality-reduction">Data visualisation using dimensionality reduction<a class="anchor" aria-label="anchor" href="#data-visualisation-using-dimensionality-reduction"></a>
</h2>
<div class="section level3">
<h3 id="data-normalisation-and-scaling">Data normalisation and scaling<a class="anchor" aria-label="anchor" href="#data-normalisation-and-scaling"></a>
</h3>
<p>In RNA-seq data, there are often genes that are expressed at extreme
magnitudes, and the variance often skews the distribution of the data.
As many statistical methods often assume an approximately normal
distribution, log normalisation of the data is a crucial step to
stabilise the variance in our data and make it more interpretable.</p>
<p>After log normalisation of the data, the next step is to find highly
variable genes (HVGs). As the feature space is large, we often perform
this step to detect crucial genes that contribute to the cell-cell
variation in scRNA-seq. Not only does this help narrow down the feature
space to include biologically informative genes, it will also help
decrease the computational time of downstream analyses.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seu</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seu</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seu</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, do.scale <span class="op">=</span> <span class="cn">TRUE</span>, do.center <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="performing-dimensionality-reduction">Performing dimensionality reduction<a class="anchor" aria-label="anchor" href="#performing-dimensionality-reduction"></a>
</h3>
<p>Principal Component Analysis (PCA) is a statistical technique to
transform high dimensionality data into low dimensionality data whilst
retaining the variability of the data into Principal Components (PCs).
It is typically performed prior to running any other dimensionality
reduction technique as algorithms such as t-SNE and UMAP are
computationally intensive, this reduces the computational load and
memory required whilst optimizing the performance.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html" class="external-link">RunTSNE</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, perplexity<span class="op">=</span><span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, min.dist <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p>We will visualise the data using the three dimensionality reduction
approaches above and color the individual data points (denoting single
cells) by their provided cell type labels. Without much knowledge of the
biological system, can you guess whether we have discrete or continuous
cell types? What can you infer about the lineages?</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, group.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"TSNE"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, group.by <span class="op">=</span> <span class="st">"cell_cluster"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"cell_cluster"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"TSNE"</span><span class="op">)</span></span>
<span><span class="va">p6</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_cluster"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span>, <span class="va">p5</span>, <span class="va">p6</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-8-1.png" width="1440"></p>
<p><strong>Exercises:</strong></p>
<ul>
<li>What happens to the visualisations when you change the TSNE
perplexity parameter?</li>
<li>What happens to the visualisations when you change the UMAP min.dist
parameter?</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html" class="external-link">RunTSNE</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, reduction.name <span class="op">=</span> <span class="st">"tsne"</span>, perplexity<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">tmp</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, reduction.name <span class="op">=</span> <span class="st">"umap"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, min.dist <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">tmp</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="unsupervised-clustering-and-annotation-of-scrna-seq-data">Unsupervised clustering and annotation of scRNA-seq data<a class="anchor" aria-label="anchor" href="#unsupervised-clustering-and-annotation-of-scrna-seq-data"></a>
</h2>
<p>There may not always be a suitable reference dataset for your
single-cell RNA-seq experiment, therefore we need to start by first
identifying how many distinct groups of populations we can find from our
data. One common method to achieve this is a statistical technique
called <em>clustering</em>. A clustering method will group similar
samples (cells) together and partition samples that are different by
comparing their feature information (gene expression).</p>
<p><br></p>
<div class="section level3">
<h3 id="how-many-clusters-do-i-have-in-my-data">How many clusters do I have in my data?<a class="anchor" aria-label="anchor" href="#how-many-clusters-do-i-have-in-my-data"></a>
</h3>
<p>We will cluster the cells using the Seurat implementation of louvain
clustering. When we separately visualise the cells by time, we see the
cells overlap regardless of which timepoint they belong to, therefore we
will cluster the data together.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, group.by <span class="op">=</span> <span class="st">"age"</span>, split.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"age"</span>, split.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"TSNE"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"age"</span>, split.by <span class="op">=</span> <span class="st">"age"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">1</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-11-1.png" width="1440"></p>
<p>Since we know the number of cell types expected, for the purpose of
this tutorial, we will set a resolution parameter that gives us a
similar number of clusters equal to unique cell types.</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What does the resolution parameter do?</li>
<li>How do we choose the right resolution?</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, resolution <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 5412</span></span>
<span><span class="co">#&gt; Number of edges: 178258</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9560</span></span>
<span><span class="co">#&gt; Number of communities: 4</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="for-a-gene-of-interest-how-can-i-visualise-the-gene-expression-distribution">For a gene of interest, how can I visualise the gene expression
distribution?<a class="anchor" aria-label="anchor" href="#for-a-gene-of-interest-how-can-i-visualise-the-gene-expression-distribution"></a>
</h3>
<p>Then, once we have generated the unique clusters, we can annotate the
data using known marker genes.</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>How do you think the unsupervised clustering performed?</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_cluster"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-13-1.png" width="1152"></p>
<p>Here are some known marker genes of the cell types we expect in the
data:</p>
<ul>
<li>nIPCs: Neurog2, Tfap2c</li>
<li>Neuroblast: Eomes, Igfbpl1</li>
<li>Immature Granule Cell (GC): Prox1, Fxyd7</li>
<li>Immature Pyramidal Cell (Pyr): Ncam1</li>
</ul>
<p>Feel free to visualise different marker genes to visualise the
distribution of the gene expression across cells.</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What parameter would you change in <code>FeaturePlot</code> to view
the data by age?</li>
<li>Do you think these marker genes are informative to identify the
specific cell types?</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TFAP2C"</span>, <span class="st">"NEUROG2"</span>, <span class="st">"EOMES"</span>, <span class="st">"IGFBPL1"</span>, <span class="st">"FXYD1"</span>, <span class="st">"PROX1"</span>, <span class="st">"NCAM1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="va">markers</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, pt.size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
<p>Alternatively, we can use the cluster assignments and perform
differential expression analysis to find differential genes between
clusters to define cell types. Lets use it to see if we can find markers
for the immature pyramidal cells.</p>
<p>As this step takes a few minutes, we can load in the pre-calculated
results.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"../data/cluster1.markers_RNA_snn_res.0.1.rda"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cluster1.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">seu</span>, ident.1 <span class="op">=</span> <span class="fl">1</span>, test.use <span class="op">=</span> <span class="st">"wilcox"</span>, min.pct<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">cluster1.markers</span>, file<span class="op">=</span><span class="st">"data/cluster1.markers_RNA_snn_res.0.1.rda"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"../data/cluster1.markers_RNA_snn_res.0.1.rda"</span><span class="op">)</span></span>
<span>  <span class="va">cluster1.markers</span> <span class="op">&lt;-</span> <span class="va">cluster1.markers_RNA_snn_res.0.1</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pos_only</span> <span class="op">&lt;-</span> <span class="va">cluster1.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What do the top markers tell you about the identity of this
cluster?</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fill in the code to visualise the new marker genes obtained from the DE analysis.</span></span></code></pre></div>
<p><strong>[Optional]</strong> We can evaluate the clustering
performance using the cell type labels provided from the
publications.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="co"># compare with a different resolution</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 5412</span></span>
<span><span class="co">#&gt; Number of edges: 178258</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8966</span></span>
<span><span class="co">#&gt; Number of communities: 11</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ARI</span></span>
<span><span class="va">ari</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span>, <span class="va">seu</span><span class="op">$</span><span class="va">RNA_snn_res.0.1</span><span class="op">)</span>,</span>
<span>        <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span>, <span class="va">seu</span><span class="op">$</span><span class="va">RNA_snn_res.0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># NMI</span></span>
<span><span class="va">nmi</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/compare.html" class="external-link">compare</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        <span class="va">seu</span><span class="op">$</span><span class="va">RNA_snn_res.0.1</span>, method <span class="op">=</span> <span class="st">"nmi"</span><span class="op">)</span>,</span>
<span>        <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/compare.html" class="external-link">compare</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        <span class="va">seu</span><span class="op">$</span><span class="va">RNA_snn_res.0.5</span>, method <span class="op">=</span> <span class="st">"nmi"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  res <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"res_0.1"</span>, <span class="st">"res_0.5"</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ari</span>, <span class="va">nmi</span><span class="op">)</span>,</span>
<span>  eval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ARI"</span>, <span class="st">"NMI"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">res</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span>col <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">eval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Clustering resolution"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Evalution score"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What are additional visualisations can we use to determine the
annotations by marker gene expression?</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="what-is-the-cell-type-composition-of-my-data">What is the cell type composition of my data?<a class="anchor" aria-label="anchor" href="#what-is-the-cell-type-composition-of-my-data"></a>
</h3>
<p>Once we have identified the cell types, often it is of interest to
visualise the cell type proportions across conditions (ie. age).</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>Why is it useful to visualise both the absolute counts and the
proportions?</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span>, <span class="va">seu</span><span class="op">$</span><span class="va">age</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Var2</span>,</span>
<span>           y <span class="op">=</span> <span class="va">Freq</span>,</span>
<span>           fill <span class="op">=</span> <span class="va">Var1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Age"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Frequency"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Composition of cell types"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Var2</span>,</span>
<span>           y <span class="op">=</span> <span class="va">Freq</span>,</span>
<span>           fill <span class="op">=</span> <span class="va">Var1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"fill"</span>, stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Age"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Frequency"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Composition of cell types"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span> <span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="perform-cell-classification-with-scclassify">Perform cell classification with <code>scClassify</code><a class="anchor" aria-label="anchor" href="#perform-cell-classification-with-scclassify"></a>
</h2>
<p>To take advantage of the large collection of well-annotated scRNA-seq
datasets, the <code>scClassify</code> package implements a set of
methods to perform accurate cell type classification based on ensemble
learning and sample size calculation. More details can be viewed from
the original <a href="https://doi.org/10.15252/msb.20199389" class="external-link">paper</a></p>
<p>The code below shows an example how to utilise a subset of data as
reference data and then classify on the remaining data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span></span>
<span><span class="co"># subsample 70% of the data as training dataset, and rest as test</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">*</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">seu</span>, layer<span class="op">=</span><span class="st">"data"</span><span class="op">)</span><span class="op">[</span>, <span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">seu</span>, layer<span class="op">=</span><span class="st">"data"</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="va">idx</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cellTypes_train</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">cellTypes_test</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cellTypes_train</span><span class="op">)</span></span>
<span><span class="co">#&gt; cellTypes_train</span></span>
<span><span class="co">#&gt;  Immature-GC Immature-Pyr   Neuroblast         nIPC </span></span>
<span><span class="co">#&gt;         1133         1938          601          116</span></span></code></pre></div>
<p>We first perform non-ensemble scClassify by using 70% data as our
reference dataset. We use <code>WKNN</code> as the KNN algorithm,
<code>DE</code> (differential expression genes) as the gene selection
method. As per our previous study, we found Pearson correlation to be
the optimal similarity metric for comparing single cell RNA-seq data (<a href="https://doi.org/10.1093/bib/bby076" class="external-link">Kim et al., 2018</a>, thus we
choose pearson as the similarity calculation method.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_scClassify</span> <span class="op">&lt;-</span> <span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/train_scClassify.html" class="external-link">train_scClassify</a></span><span class="op">(</span>exprsMat_train <span class="op">=</span> <span class="va">train</span>, </span>
<span>                                           cellTypes_train <span class="op">=</span> <span class="va">cellTypes_train</span>,</span>
<span>                                           algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                                           selectFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                                           similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>                                           returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                           verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; after filtering not expressed genes </span></span>
<span><span class="co">#&gt; [1] 9990 3788</span></span>
<span><span class="co">#&gt; [1] "Feature Selection..."</span></span>
<span><span class="co">#&gt; [1] "Number of genes selected to construct HOPACH tree 165"</span></span>
<span><span class="co">#&gt; [1] "Constructing tree ..."</span></span>
<span><span class="co">#&gt; [1] "Training...."</span></span>
<span><span class="co">#&gt; [1] "=== selecting features by: limma ===="</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">train_scClassify</span></span>
<span><span class="co">#&gt; Class: scClassifyTrainModel </span></span>
<span><span class="co">#&gt; Model name: training </span></span>
<span><span class="co">#&gt; Feature selection methods: limma </span></span>
<span><span class="co">#&gt; Number of cells in the training data: 3788 </span></span>
<span><span class="co">#&gt; Number of cell types in the training data: 4</span></span></code></pre></div>
<p>We can visualise the cell type hierarchy tree using
<code>plotCellTypeTree()</code>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/plotCellTypeTree.html" class="external-link">plotCellTypeTree</a></span><span class="op">(</span><span class="va">train_scClassify</span><span class="op">@</span><span class="va">cellTypeTree</span>, group_level <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>Next, we perform <code>predict_scClassify()</code> with our trained
model <code>trainRes = train_scClassify</code> to predict the cell types
of our query data matrix <code>test</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_res</span> <span class="op">&lt;-</span> <span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/predict_scClassify.html" class="external-link">predict_scClassify</a></span><span class="op">(</span>exprsMat_test <span class="op">=</span> <span class="va">test</span>,</span>
<span>                               cellTypes_test <span class="op">=</span> <span class="va">cellTypes_test</span>,</span>
<span>                               trainRes <span class="op">=</span> <span class="va">train_scClassify</span>,</span>
<span>                               algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                               similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>                               prob_threshold <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Ensemble learning is disabled... </span></span>
<span><span class="co">#&gt; Using parameters: </span></span>
<span><span class="co">#&gt; similarity  algorithm   features </span></span>
<span><span class="co">#&gt;  "pearson"     "WKNN"    "limma" </span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; classify_res</span></span>
<span><span class="co">#&gt;                correct   correctly unassigned           intermediate </span></span>
<span><span class="co">#&gt;             0.93103448             0.00000000             0.01046798 </span></span>
<span><span class="co">#&gt; incorrectly unassigned         error assigned          misclassified </span></span>
<span><span class="co">#&gt;             0.02709360             0.00000000             0.03140394 </span></span>
<span><span class="co">#&gt; weights for each base method: </span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">scClassify_res</span> <span class="op">&lt;-</span> <span class="va">pred_res</span><span class="op">$</span><span class="va">pearson_WKNN_limma</span><span class="op">$</span><span class="va">predRes</span></span></code></pre></div>
<p>Then, we check prediction results with the original labels.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_scClassify</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu</span>, reduction<span class="op">=</span><span class="st">"umap"</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                           label <span class="op">=</span> <span class="va">scClassify_res</span>,</span>
<span>                           method <span class="op">=</span> <span class="st">"scClassify"</span><span class="op">)</span></span>
<span><span class="va">df_GT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu</span>, reduction<span class="op">=</span><span class="st">"umap"</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                           label <span class="op">=</span> <span class="va">seu</span><span class="op">$</span><span class="va">cell_cluster</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">]</span>,</span>
<span>                           method <span class="op">=</span> <span class="st">"Ground Truth"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plotdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_scClassify</span>, <span class="va">df_GT</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plotdf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, colour <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"UMAP plot"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basics_of_single-cell_annotation_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>We can quickly quantify the number of cells assigned to each unique
label.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">scClassify_res</span>, <span class="va">cellTypes_test</span><span class="op">)</span></span>
<span><span class="co">#&gt;                           cellTypes_test</span></span>
<span><span class="co">#&gt; scClassify_res             Immature-GC Immature-Pyr Neuroblast nIPC</span></span>
<span><span class="co">#&gt;   Immature-GC                      476            3         34    0</span></span>
<span><span class="co">#&gt;   Immature-GC_Immature-Pyr           4            0          3    0</span></span>
<span><span class="co">#&gt;   Immature-Pyr                       3          813          6    0</span></span>
<span><span class="co">#&gt;   Neuroblast                         2            0        169    0</span></span>
<span><span class="co">#&gt;   Neuroblast_nIPC                    0            0         12    1</span></span>
<span><span class="co">#&gt;   nIPC                               0            0          0   54</span></span>
<span><span class="co">#&gt;   unassigned                        10            3         31    0</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="save-the-processed-data">Save the processed data<a class="anchor" aria-label="anchor" href="#save-the-processed-data"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reducedDim(sce, "PCA") &lt;- Embeddings(seu, reduction="pca")</span></span>
<span><span class="co"># reducedDim(sce, "TSNE") &lt;- Embeddings(seu, reduction="tsne")</span></span>
<span><span class="co"># reducedDim(sce, "UMAP") &lt;- Embeddings(seu, reduction="umap")</span></span>
<span></span>
<span><span class="co"># save(sce, file="data/hippocampus_minimal.rda")</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="references-and-other-useful-resources">References and other useful resources<a class="anchor" aria-label="anchor" href="#references-and-other-useful-resources"></a>
</h2>
<ul>
<li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat
clustering tutorial</a></li>
</ul>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carissa Chen, Sharon Long, Pengyi Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
