<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SCMultiomeCNRSR">
<title>Introduction to 10x Multiome • SCMultiomeCNRSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to 10x Multiome">
<meta property="og:description" content="SCMultiomeCNRSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCMultiomeCNRSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction_to_single-cell.html">Introduction to single-cell pre-processing</a>
    <a class="dropdown-item" href="../articles/basics_of_single-cell_annotation.html">Basics of single-cell annotation</a>
    <a class="dropdown-item" href="../articles/advancedphenotyping.html">Advanced cell phenotyping</a>
    <a class="dropdown-item" href="../articles/citefuse.html">Analysing CITE-seq data with CiteFuse</a>
    <a class="dropdown-item" href="../articles/introduction_to_10x_multiome.html">Introduction to 10x multiome</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to 10x Multiome</h1>
                        <h4 data-toc-skip class="author">Siqu Long<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:siqu.long@sydney.edu.au" class="email"&gt;siqu.long@sydney.edu.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Carissa
Chen<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:cchen@uni.sydney.edu.au" class="email"&gt;cchen@uni.sydney.edu.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
Pengyi Yang<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:pengyi.yang@sydney.edu.au" class="email"&gt;pengyi.yang@sydney.edu.au&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R/blob/HEAD/vignettes/introduction_to_10x_multiome.Rmd" class="external-link"><code>vignettes/introduction_to_10x_multiome.Rmd</code></a></small>
      <div class="d-none name"><code>introduction_to_10x_multiome.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Single-cell RNA sequencing is an effective method for analyzing the
characteristics and dynamics in tissues in both healthy and diseased
states. Nonetheless, to gain a comprehensive understanding of cellular
conditions during illness or treatment, it can be beneficial to collect
information beyond just the gene expression profile of a cell and turns
to multimodal single-cell analysis. Single-cell multi-omics technologies
concurrently combines different single-modality omics methods that
profile the transcriptome, genome, epigenome, epitranscriptome,
proteome, metabolome, and other (emerging) omics <span class="citation">(Baysoy et al. 2023)</span>, which provides a
comprehensive view and detailed characterization of individual cell
states and activities.</p>
<p>In our previous tutorial, we learned the workflow of multimodal
single-cell omics analysis with the CITE-seq data, which combines
scRNA-seq with antibody-derived tags to measure protein levels on the
cell surface. In this tutorial, let’s move to another commonly applied
multiome data that is derived from simultaneous profiling of the
transcriptome (scRNA-seq) and chromatin accessibility (scATAC-seq).</p>
<div class="section level3">
<h3 id="preparation-and-assumed-knowledge">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h3>
<ul>
<li>Knowledge of R syntax <a href="https://www.w3schools.com/r/default.asp" class="external-link">Basic tutorials of
R</a>
</li>
<li>Basic knowledge in <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">single
cell data analysis</a>
</li>
<li>Familiarity with the <a href="https://satijalab.org/seurat/articles/get_started_v5_new" class="external-link">Seurat
class</a> and <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html" class="external-link">SingleCellExperiment
class</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Get familiar with the concepts of RNA + ATAC multiomic sequencing
and analysis</li>
<li>Understand the 10x multiome data</li>
<li>Learn about the common practices of preprocessing the RNA + ATAC
multiomic data, using the 10x multiome sample data</li>
<li>Prepare the RNA + ATAC data for multiomic integrative analysis using
Matilda</li>
<li>Integrative analysis of the RNA + ATAC multiomic data using
Matilda</li>
</ul>
<p><br><br></p>
</div>
<div class="section level3">
<h3 id="schedule">Schedule<a class="anchor" aria-label="anchor" href="#schedule"></a>
</h3>
<p>Structure of this tutorial:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1 Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>2 Hello, 10x Multiome!</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>3 Preparation of required data files</td>
<td>10m</td>
</tr>
<tr class="even">
<td>4 Data preprocessing and analysis</td>
<td>30m</td>
</tr>
<tr class="odd">
<td>5 Integrative analysis with Matilda</td>
<td>60m</td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="hello-10x-multiome">Hello, 10x Multiome!<a class="anchor" aria-label="anchor" href="#hello-10x-multiome"></a>
</h2>
<p>The <strong>10x Genomics’ Single Cell Multiome ATAC + Gene Expression
solution (10x Multiome)</strong> is one of the dominant technology for
combining the scRNA-seq and scATAC-seq. It integrates single-cell RNA
sequencing (scRNA-seq) with a single-cell epigenetic readout in the form
of chromatin accessibility profiling with single-cell ATAC-seq
(scATAC-seq). With the joint profiling of RNA and ATAC, it connects
information on the activity of regulatory elements to gene expression
data in the same nucleus, which establishes a direct one-to-one
connection between gene expression and epigenetic programs that are not
capturable otherwise.</p>
<p>Figure 1 provides an overview of the 10x Multiome workflow. In
summary, Nuclei isolation is followed by “tagmentation” with Tn5
transposase, which fragments open chromatin and tags DNA with adapters
for 10x Barcodes. The nuclei are then added to the 10x Chromium X, where
specialized Next GEM beads capture adapters and mRNAs, adding cell- and
molecule-specific barcodes. The mixture is then extracted, pooled, and
undergoes library preparation, including amplification and cleanup,
which can be then used for the next-generation sequencing. More details
with step-by-step explanation can be found in the optional reading
below.</p>
<div class="figure" style="text-align: center">
<img src="figures/workflow.png" alt="Figure 1. 10x Multiome workflow. The paired ATAC and gene expression libraries are generatred from the isolated nuclei, which are then sequenced. (Source: 10x Genomics)" width="98%"><p class="caption">
Figure 1. 10x Multiome workflow. The paired ATAC and gene expression
libraries are generatred from the isolated nuclei, which are then
sequenced. (Source: 10x Genomics)
</p>
</div>
<details><summary><strong><em>(Optional Reading) How is the 10x multiome (RNA + ATAC) data
profiled and prepared exactly?</em></strong>
</summary><p><br> Specifically, the 10x Multiome combines the pre-existing
scRNA-seq with a scATAC-seq technology, namely 10x Genomics’ (3’)
Chromium Single Cell Gene Expression (<a href="https://www.scdiscoveries.com/blog/knowledge/single-cell-transcriptomics/#how-it-works" class="external-link">Click
to see the workflow</a>), and Chromium Single Cell ATAC (<a href="https://www.scdiscoveries.com/blog/knowledge/how-scatac-seq-works/#the-method" class="external-link">Click
to see the workflow</a>). According to the <a href="https://assets.ctfassets.net/an68im79xiti/7x5E4P6xefQruTbFg0yr3a/1381fdcd2d2e7d667ef5b415119dab15/CG000338_ChromiumNextGEM_Multiome_ATAC_GEX_User_Guide_RevE.pdf#page=32.10" class="external-link">user
guide for the 10x Multiome</a>, it mainly consists of the following 8
steps to acquire the scRNA+scATAC data:</p>
<p><strong>Step 1. Transposition</strong> Nuclei suspensions are
incubated in a Transposition Mix that includes a Transposase. The
Transposase enters the nuclei and preferentially fragments the DNA in
open regions of the chromatin. Simultaneously, adapter sequences are
added to the ends of the DNA fragments.</p>
<p><strong>Step 2. GEM Generation &amp; Barcoding</strong> Similar to
the aforementioned 10x Genomics’ (3’) Chromium Single Cell Gene
Expression and Chromium Single Cell ATAC workflows, 10x Multiome also
relies on the next GEMs (Gel bead-in-EMulsion) in the Chromium X. As is
shown in the Figure 2 below, in the 10x Multiome protocol, the Gel bead
include (1) a poly(dT) sequence that enables production of barcoded,
full-length cDNA from poly-adenylated mRNA for gene expression (GEX)
library, and (2) a Spacer sequence that enables barcode attachment to
transposed DNA fragments for ATAC library.</p>
<div class="figure" style="text-align: center">
<img src="figures/10x_gem_multi.png" alt="Figure 2. Two types of oligos line a 10x Multiome bead. One contains a poly(dT)VN to bind Poly(A) tails of mRNA transcripts, another a spacer that binds the Tn5-adapted DNA fragments of the scATAC library. (Source: 10x Genomics)" width="70%"><p class="caption">
Figure 2. Two types of oligos line a 10x Multiome bead. One contains a
poly(dT)VN to bind Poly(A) tails of mRNA transcripts, another a spacer
that binds the Tn5-adapted DNA fragments of the scATAC library. (Source:
10x Genomics)
</p>
</div>
<p>GEMs are generated by combining barcoded Gel Beads, transposed
nuclei, a Master Mix, and Partitioning Oil on a Chromium Next GEM Chip J
(See Figure 3 below).</p>
<div class="figure" style="text-align: center">
<img src="figures/gem.png" alt="Figure 3. GEM generation protocol. To achieve single nuclei resolution, the nuclei are delivered at a limiting dilution, such that the majority (~90-99%) of generated GEMs contains no nuclei, while the remainder largely contain a single nucleus. (Source: 10x Genomics)" width="50%"><p class="caption">
Figure 3. GEM generation protocol. To achieve single nuclei resolution,
the nuclei are delivered at a limiting dilution, such that the majority
(~90-99%) of generated GEMs contains no nuclei, while the remainder
largely contain a single nucleus. (Source: 10x Genomics)
</p>
</div>
<p>Upon GEM generation, the Gel Bead dissolves, releasing
oligonucleotides containing an Illumina® P5 sequence, a 16 nt 10x
Barcode (for ATAC), and a Spacer sequence. In the same partition,
primers with an Illumina® TruSeq Read 1, a 16 nt 10x Barcode (for GEX),
a 12 nt unique molecular identifier (UMI), and a 30 nt poly(dT) sequence
are also released. These primers mix with the nuclei lysate containing
transposed DNA fragments, mRNA, and Master Mix, which includes reverse
transcription (RT) reagents.</p>
<p>Incubation of the GEMs, as is shown in Figure 4 below, produces 10x
Barcoded DNA from the transposed DNA (for ATAC) and 10x Barcoded,
full-length cDNA from poly-adenylated mRNA (for GEX). This is followed
by a quenching step that stops the reaction.</p>
<div class="figure" style="text-align: center">
<img src="figures/insidegem.png" alt="Figure 4. What is happening during the incubation inside individual GEMs. (Source: 10x Genomics)" width="90%"><p class="caption">
Figure 4. What is happening during the incubation inside individual
GEMs. (Source: 10x Genomics)
</p>
</div>
<p><strong>Step 3. Post-GEM Cleanup</strong> After the GEMs incubation,
the GEMs are broken and pooled fractions are recovered. Silane magnetic
beads are used to purify the cell barcoded products from the post GEM-RT
reaction mixture, which includes leftover biochemical reagents and
primers.</p>
<p><strong>Step 4. Pre-Amplification PCR</strong> Barcoded transposed
DNA and barcoded full length cDNA from poly-adenylated mRNA are
amplified via PCR to fill gaps and for generating sufficient mass for
library construction (See Figure 5). The pre-amplified product is used
as input for both ATAC library construction and cDNA amplification for
gene expression library construction.</p>
<div class="figure" style="text-align: center">
<img src="figures/amplification.png" alt="Figure 5. Pooled pre-amplification PCR. (Source: 10x Genomics)" width="70%"><p class="caption">
Figure 5. Pooled pre-amplification PCR. (Source: 10x Genomics)
</p>
</div>
<p><strong>Step 5. ATAC Library Construction</strong> P7 and a sample
index are added to pre-amplified transposed DNA during ATAC library
construction via PCR. The final ATAC libraries contain the P5 and P7
sequences used in Illumina® bridge amplification (See Figure 6).</p>
<div class="figure" style="text-align: center">
<img src="figures/ataclib.png" alt="Figure 6. ATAC library construction. (Source: 10x Genomics)" width="50%"><p class="caption">
Figure 6. ATAC library construction. (Source: 10x Genomics)
</p>
</div>
<p><strong>Step 6. GEM cDNA Aplification</strong> Barcoded, full-length
pre-amplified cDNA is amplified via PCR to generate sufficient mass for
gene expression library construction.</p>
<p><strong>Step 7. Gene Expression Library Construction</strong>
Enzymatic fragmentation and size selection are used to optimize the cDNA
amplicon size. P5, P7, i7 and i5 sample indexes, and TruSeq Read 2 (read
2 primer sequence) are added via End Repair, A-tailing, Adaptor
Ligation, and PCR. The final gene expression libraries contain the P5
and P7 primers used in Illumina® bridge amplification (See Figure
7).</p>
<div class="figure" style="text-align: center">
<img src="figures/rnalib.png" alt="Figure 7. cDNA amplification &amp; gene expression library construction. (Source: 10x Genomics)" width="50%"><p class="caption">
Figure 7. cDNA amplification &amp; gene expression library construction.
(Source: 10x Genomics)
</p>
</div>
<p><strong>Step 8. Sequencing</strong> As is demonstrated in Figure 8,
Chromium Single Cell Multiome ATAC libraries comprise double stranded
DNA with standard Illumina® paired-end constructs which begin with P5
and end with P7. Sequencing these libraries produces a standard
Illumina® BCL data output folder that includes paired-end Read 1N and
Read 2N used for sequencing the DNA insert, along with the 8 bp sample
index in the i7 read and 16 bp 10x Barcode sequence in the i5 read .</p>
<div class="figure" style="text-align: center">
<img src="figures/seq1.png" alt="Figure 8. Chromium Single Cell Multiome ATAC library.  (Source: 10x Genomics)" width="70%"><p class="caption">
Figure 8. Chromium Single Cell Multiome ATAC library. (Source: 10x
Genomics)
</p>
</div>
<p>As is illustrated in Figure 9, Chromium Single Cell Multiome Gene
Expression libraries comprise cDNA insert with standard Illumina®
paired-end constructs which begin with P5 and end with P7. Sequencing
these libraries produces a standard Illumina® BCL data output folder.
TruSeq Read 1 is used to sequence 16 bp 10x Barcodes and 12 bp UMI,
while 10 bp i5 and i7 sample index sequences are the sample index reads.
TruSeq Read 2 is used to sequence the insert.</p>
<div class="figure" style="text-align: center">
<img src="figures/seq2.png" alt="Figure 9. Chromium Single Cell Multiome Gene Expression library. (Source: 10x Genomics)" width="70%"><p class="caption">
Figure 9. Chromium Single Cell Multiome Gene Expression library.
(Source: 10x Genomics)
</p>
</div>
</details><p><br><br> After sequencing, the sequencer outputs are processed and
pre-analyzed by the <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/what-is-cell-ranger-arc" class="external-link">Cell
Ranger ARC pipelines</a> from 10x Genomics. In the following sections,
we will learn how to load the RNA and ATAC data from the serious output
files from Cell Ranger.</p>
</div>
<div class="section level2">
<h2 id="preparation-of-required-data-files">Preparation of required data files<a class="anchor" aria-label="anchor" href="#preparation-of-required-data-files"></a>
</h2>
<p>In this section, we first prepare our sample 10x multiome data. We
will use the PBMC multiome dataset publicly available from <a href="https://www.10xgenomics.com/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-1-0-0" class="external-link">10x
genomics</a>. This PBMCs is from a healthy female donor aged 25 and were
obtained by 10x Genomics from AllCells. In this dataset, scRNA-seq and
scATAC-seq profiles were simultaneously collected in the same cells.
From the <a href="https://www.10xgenomics.com/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-1-0-0" class="external-link">dataset
introduction site</a>, we can see that a series of output files from the
Cell Ranger are available (See Figure 10 below). A detailed introduction
for each of the file can be found from the <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/output/overview" class="external-link">Cell
Ranger guide</a>.</p>
<div class="figure" style="text-align: center">
<img src="figures/outputs.png" alt="Figure 10. Output files of our sample data available for download. (Source: 10x Genomics)" width="70%"><p class="caption">
Figure 10. Output files of our sample data available for download.
(Source: 10x Genomics)
</p>
</div>
<p>Here, our goal is to derive the RNA and ATAC matrix. We will only
focus on the following three files:</p>
<ul>
<li><p><code>pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5</code>:
The filtered feature barcode matrix (HDF5). The rows represent features,
genes and peaks detected, while the columns consist of all
cell-associated barcodes.</p></li>
<li><p><code>pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz</code>:
The ATAC Per fragment information file (TSV.GZ). It contains one line
per unique fragment with tab-separated fields, including the reference
genome chromosome of fragment (<code>chrom</code>), the adjusted start
and end position of fragment on chromosome (<code>chromStart</code>,
<code>chromEnd</code>), the 10x barcode of the fragment
(<code>barcode</code>), and the total number of read pairs associated
with this fragment (<code>readSupport</code>).</p></li>
<li><p><code>pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz.tbi</code>:
The ATAC Per fragment information index (TSV.GZ index). It is a <a href="http://www.htslib.org/doc/tabix.html" class="external-link">tabix</a> index of the
fragment intervals facilitating random access to records from an
arbitrary genomic interval.</p></li>
</ul>
<p>To facilitate easy loading and exploration, we pre-downloaded the
three files already. You can find them in the <code>data</code> folder.
However, you can also download these required data by yourself later via
running the following lines in a shell.</p>
<details><summary><strong><em>(Optional) Downloading the data yourself</em></strong>
</summary><div class="sourceCode" id="cb1"><pre class="sourceCode bash fold-show"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_3k/pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_3k/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_3k/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz.tbi</span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="data-preprocessing-and-analysis">Data preprocessing and analysis<a class="anchor" aria-label="anchor" href="#data-preprocessing-and-analysis"></a>
</h2>
<p>Now we have data files ready! In the following sections, we will
prepare the matrix data in R and conduct some essential data
preprocessing and analysis. We will mostly follow the common practice of
joint RNA and ATAC analysis introduced by <a href="https://stuartlab.org/signac/articles/pbmc_multiomic" class="external-link">Signac</a>.</p>
<div class="section level3">
<h3 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<p>Before getting start, let us load some of the R libraries we will
use.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span> <span class="co"># get gene annotations for hg19 for creating the ATAC assay</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mojaveazure.github.io/seurat-disk/" class="external-link">SeuratDisk</a></span><span class="op">)</span> <span class="co"># for load the annotated pbmc reference dataset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h3>
<p>With the three required files, we can easily create a Seurat object
based on the gene expression data, and then add in the ATAC-seq data as
a second assay. First, let us load the read count matrix from 10X
CellRanger hdf5 file into R, using the <code>Read10x_H5</code> function
from <code>Seurat</code>. This will give us the access to both RNA data
(<code>Gene Expression</code>) and ATAC data matrix
(<code>Peaks</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># This will give us both the RNA and ATAC matrix data, named as 'Gene Expression' and 'Peaks' respectively.</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"../data/pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span></code></pre></div>
<p><br> Now, we can create a <code>Seurat</code> object named
<code>pbmc</code> and initialize it with the RNA data matrix.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a Seurat object containing the Gene Expression matrix</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">`Gene Expression`</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br> We are already familiar with the RNA cell by feature matrix
(<code>Gene Expression</code>). Let’s have a look at how the ATAC data
matrix (<code>Peaks</code>) looks like. As can be seen in the first 10
rows by first 10 columns of the <code>Peaks</code> matrix below, it is a
sparse matrix with non-zero values stored only. It looks much like the
scRNA-seq cell by feature count matrix. The major difference is the
features (row names) are not genes but the regions of the peaks that are
predicted to be open chromatin by the peak callers. The non-zero values
stored in matrix represents the number of tn5 integration site in each
of the cell (column) that maps to each region (row).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span><span class="op">$</span><span class="va">Peaks</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt; 10 x 10 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;                                       </span></span>
<span><span class="co">#&gt; chr1:9986-10403    . . . . . . . . 2 .</span></span>
<span><span class="co">#&gt; chr1:180690-181695 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:191335-192015 . . . . . . . 2 . .</span></span>
<span><span class="co">#&gt; chr1:267775-268224 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:271144-271510 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:586016-586386 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:629708-630181 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:633785-634272 . . . . . . . . . .</span></span>
<span><span class="co">#&gt; chr1:777438-779964 . . 2 2 4 . . 2 2 .</span></span>
<span><span class="co">#&gt; chr1:816129-817671 . 2 . . . . . . . .</span></span></code></pre></div>
<p><br> For creating the chromatin assay object, we also need
information from the other two files: the ATAC Per fragment information
file and the ATAC Per fragment information index. We will store the
created ATAC assay in the <code>pbmc</code> object as a secondary assay.
Note that in the following code, we also add gene annotations to the
ATAC assay object for the human genome. This allows downstream functions
to pull the gene annotation information directly from the object.</p>
<details><summary><strong><em>What does the fragment file look like?</em></strong>
</summary><p><br> As is introduced above, the ATAC Per fragment information file
(TSV.GZ) contains one line per unique fragment with tab-separated
fields, including the reference genome chromosome of fragment
(<code>chrom</code>), the adjusted start and end position of fragment on
chromosome (<code>chromStart</code>, <code>chromEnd</code>), the 10x
barcode of the fragment (<code>barcode</code>), and the total number of
read pairs associated with this fragment (<code>readSupport</code>).
Let’s read the fragment file and see what do these values actually look
like:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read file (only read 10 rows)</span></span>
<span><span class="va">fragpath</span> <span class="op">&lt;-</span> <span class="st">"../data/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz"</span></span>
<span><span class="va">frag.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">fragpath</span>, header<span class="op">=</span><span class="cn">F</span>, nrows <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># change the column names </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">frag.file</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chrom'</span>,<span class="st">'chromStart'</span>,<span class="st">'chromEnd'</span>,<span class="st">'barcode'</span>,<span class="st">'readSupport'</span><span class="op">)</span></span>
<span><span class="co"># preview of the file content</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">frag.file</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chrom chromStart chromEnd            barcode readSupport</span></span>
<span><span class="co">#&gt; 1  chr1      10073    10209 TTTAGCAAGGTAGCTT-1           1</span></span>
<span><span class="co">#&gt; 2  chr1      10079    10333 AGCCGGTTCCGGAACC-1           2</span></span>
<span><span class="co">#&gt; 3  chr1      10096    10344 CGTTAGGTCATTAGTG-1           1</span></span>
<span><span class="co">#&gt; 4  chr1      10151    10180 AAACCGCGTGAGGTAG-1           2</span></span>
<span><span class="co">#&gt; 5  chr1      10151    10195 AAGCCTCCACACTAAT-1           1</span></span>
<span><span class="co">#&gt; 6  chr1      10151    10202 CCGTTGCGTTAGAGCC-1           2</span></span></code></pre></div>
</details><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specifying the path to the fragment information file</span></span>
<span><span class="va">fragpath</span> <span class="op">&lt;-</span> <span class="st">"../data/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz"</span></span>
<span></span>
<span><span class="co"># convert EnsDb (Ensembl database) to GRanges object </span></span>
<span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/GetGRangesFromEnsDb.html" class="external-link">GetGRangesFromEnsDb</a></span><span class="op">(</span>ensdb <span class="op">=</span> <span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="co"># convert to UCSC style by adding the 'chr' prefix</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create ATAC assay from the 'Peaks' matrix and add it to the object, together with the gene annotation</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html" class="external-link">CreateChromatinAssay</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">Peaks</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>  fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">annotation</span> <span class="co"># store the gene annotation information</span></span>
<span><span class="op">)</span></span></code></pre></div>
<br><details><summary><strong><em>What does the annotation look like?</em></strong>
</summary><p><br> As can be seen from the output below, the
<code>annotation</code> object extracts the Exon and the coding region
information from the transcripts across the genes across all the
chromosomes along with the strand and the genomic location.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation</span></span>
<span><span class="co">#&gt; GRanges object with 3021151 ranges and 5 metadata columns:</span></span>
<span><span class="co">#&gt;                   seqnames        ranges strand |           tx_id   gene_name</span></span>
<span><span class="co">#&gt;                      &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   ENSE00001489430     chrX 276322-276394      + | ENST00000399012      PLCXD1</span></span>
<span><span class="co">#&gt;   ENSE00001536003     chrX 276324-276394      + | ENST00000484611      PLCXD1</span></span>
<span><span class="co">#&gt;   ENSE00002160563     chrX 276353-276394      + | ENST00000430923      PLCXD1</span></span>
<span><span class="co">#&gt;   ENSE00001750899     chrX 281055-281121      + | ENST00000445062      PLCXD1</span></span>
<span><span class="co">#&gt;   ENSE00001489388     chrX 281192-281684      + | ENST00000381657      PLCXD1</span></span>
<span><span class="co">#&gt;               ...      ...           ...    ... .             ...         ...</span></span>
<span><span class="co">#&gt;   ENST00000361739    chrMT     7586-8269      + | ENST00000361739      MT-CO2</span></span>
<span><span class="co">#&gt;   ENST00000361789    chrMT   14747-15887      + | ENST00000361789      MT-CYB</span></span>
<span><span class="co">#&gt;   ENST00000361851    chrMT     8366-8572      + | ENST00000361851     MT-ATP8</span></span>
<span><span class="co">#&gt;   ENST00000361899    chrMT     8527-9207      + | ENST00000361899     MT-ATP6</span></span>
<span><span class="co">#&gt;   ENST00000362079    chrMT     9207-9990      + | ENST00000362079      MT-CO3</span></span>
<span><span class="co">#&gt;                           gene_id   gene_biotype     type</span></span>
<span><span class="co">#&gt;                       &lt;character&gt;    &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt;   ENSE00001489430 ENSG00000182378 protein_coding     exon</span></span>
<span><span class="co">#&gt;   ENSE00001536003 ENSG00000182378 protein_coding     exon</span></span>
<span><span class="co">#&gt;   ENSE00002160563 ENSG00000182378 protein_coding     exon</span></span>
<span><span class="co">#&gt;   ENSE00001750899 ENSG00000182378 protein_coding     exon</span></span>
<span><span class="co">#&gt;   ENSE00001489388 ENSG00000182378 protein_coding     exon</span></span>
<span><span class="co">#&gt;               ...             ...            ...      ...</span></span>
<span><span class="co">#&gt;   ENST00000361739 ENSG00000198712 protein_coding      cds</span></span>
<span><span class="co">#&gt;   ENST00000361789 ENSG00000198727 protein_coding      cds</span></span>
<span><span class="co">#&gt;   ENST00000361851 ENSG00000228253 protein_coding      cds</span></span>
<span><span class="co">#&gt;   ENST00000361899 ENSG00000198899 protein_coding      cds</span></span>
<span><span class="co">#&gt;   ENST00000362079 ENSG00000198938 protein_coding      cds</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 25 sequences (1 circular) from GRCh38 genome</span></span></code></pre></div>
</details><p>Specifically, the <code>ChromatinAssay</code> used above for storing
the ATAC data enables some specialized functions for analysing genomic
single-cell assays such as scATAC-seq. By printing the assay we can see
some of the additional information that can be contained in the
<code>ChromatinAssay</code>, including motif information, gene
annotations, and genome information. The detailed instruction on how to
access these information stored in a <code>ChromatinAassay</code> object
can be found from <a href="https://stuartlab.org/signac/articles/data_structures#the-chromatinassay-class" class="external-link">Signac</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; ChromatinAssay data with 156607 features for 2714 cells</span></span>
<span><span class="co">#&gt; Variable features: 0 </span></span>
<span><span class="co">#&gt; Genome: </span></span>
<span><span class="co">#&gt; Annotation present: TRUE </span></span>
<span><span class="co">#&gt; Motifs present: FALSE </span></span>
<span><span class="co">#&gt; Fragment files: 1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>As is done in our previous tutorials for scRNA-seq, we need to
perform the quality control by assessing the quality of the data and
filtering the poor quality cells. Here, we will compute and look at the
per-cell quality control metrics using the DNA accessibility data and
remove cells that are outliers for these metrics, as well as cells with
low or unusually high counts for either the RNA or ATAC assay. The <a href="https://stuartlab.org/signac/reference/nucleosomesignal" class="external-link">NucleosomeSignal</a>
function below calculates the strength of the nucleosome signal per
cell. It computes the ratio of fragments between 147 bp and 294 bp
(mononucleosome) to fragments &lt; 147 bp (nucleosome-free). The <a href="https://stuartlab.org/signac/reference/tssenrichment" class="external-link">TSSEnrichment</a>
function computes the transcription start site (TSS) enrichment score
for each cell.</p>
<br><details><summary><strong><em>Question: What are the nucleosome signal and transcription
start site (TSS) enrichment score? How to utilize it to control the
quality of the DNA accessibility data? </em></strong>
</summary><p> <strong><em>Nucleosome signal</em></strong> calculates the
ratio of mononucleosome (fragments with length between 147 bp and 294
bp) to nucleosome-free regions (fragments with length less than 147 bp)
by looking at the distribution of DNA fragment lengths from the cell.
Considering the aim of scATAC-seq, we essentially want to retain only
those fragments that are not bound by the mononucleosomes, which are
free of nucleosomes and open and accessible for the other DNA binding
proteins to bind to these DNA. Thus, cells with lower nucleosome signal
indicates that it has enrichment of fragments that are not bound by
nucleosomes.</p>
<p><strong><em>Transcription start site (TSS) enrichment
score</em></strong> The transcription start site (TSS) enrichment score
was originally defined by the ENCODE consortium <span class="citation">(Consortium et al. 2012)</span> as a signal-to-noise
metric for ATAC-seq experiments. The idea is that if these fragments are
from open chromatin region that is not bound by any nucleosome, the
genes in this region will be actively transcribed or regulated.
Therefore, these nucleosome-free regions should be expected to be
enriched around the transcription start site of these genes. The
enrichment score is calculated by taking a ratio of the fragments
centered at the transcription start site over the fragments in the TSS
flanking regions that is around the TSS regions. With regards to the
scATAC-seq, we ideally want to retain those cells that have a high
transcription start site enrichment score.</p>
<p>We can see from the above that the nucleosome signal and
transcription start site enrichment score are two complementary metrics
for measuring the amount of potential nucleosome-free fragments
contained by the cells in the scATAC-seq data.</p>
</details><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span></span>
<span><span class="co"># Calculate the strength of the nucleosome signal per cell</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/NucleosomeSignal.html" class="external-link">NucleosomeSignal</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/TSSEnrichment.html" class="external-link">TSSEnrichment</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co"># We can see where these calculated results are stored in our meta.data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         </span></span>
<span><span class="co">#&gt;  [4] "nCount_ATAC"           "nFeature_ATAC"         "nucleosome_signal"    </span></span>
<span><span class="co">#&gt;  [7] "nucleosome_percentile" "TSS.enrichment"        "TSS.percentile"       </span></span>
<span><span class="co">#&gt; [10] NA</span></span></code></pre></div>
<p><br> The relationship between variables stored in the object
metadata can be visualized using the <a href="https://stuartlab.org/signac/reference/densityscatter" class="external-link">DensityScatter</a>
function from Signac. This can also be used to quickly find suitable
cutoff values for different QC metrics by setting
<code>quantiles=TRUE</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://stuartlab.org/signac/reference/DensityScatter.html" class="external-link">DensityScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, x <span class="op">=</span> <span class="st">'nCount_ATAC'</span>, y <span class="op">=</span> <span class="st">'TSS.enrichment'</span>, log_x <span class="op">=</span> <span class="cn">TRUE</span>, quantiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p><br> For the scatter plot generated above, the x-axis represents the
number of counts whereas the y-axis is the TSS enrichment score. Each
cell is colored by the density of points. We can see there are some red
lines going across the plots both vertically and horizontally. The
vertical red lines are the 5th quantile, 10th quantile, 90th quantile
and 95th quantile of the <code>nCount</code> while their corresponding
values can be found at the top of the plot. Similarly, the horizontal
red lines are the quantiles for the TSS enrichment score. By visualizing
data in this way, it helps us come up with filtering thresholds to
ensure that we are not choosing a threshold that is too stringent that
majority of our cells are filtered out and at the same time selecting a
threshold that retains high quality of cells.</p>
<p>Similarly, we can create the scatter plot with the nucleosome signal
as well.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://stuartlab.org/signac/reference/DensityScatter.html" class="external-link">DensityScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, x <span class="op">=</span> <span class="st">'nCount_ATAC'</span>, y <span class="op">=</span> <span class="st">'nucleosome_signal'</span>, log_x <span class="op">=</span> <span class="cn">TRUE</span>, quantiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p><br> Also, we can plot the distribution of each QC metric in violin
plot using the <a href="https://satijalab.org/seurat/reference/vlnplot" class="external-link">VlnPlot</a>
function from Seurat. We can simply choosing which features (metrics) to
be plotted by changing the input to the <code>features</code> parameter.
In the example below, we will plot the distribution of RNA read counts
(<code>nCount_RNA</code>), the ATAC peak counts
(<code>nCount_ATAC</code>), the TSS enrichment score
(<code>TSS.enrichment</code>) and the nucleosome signal
(<code>nucleosome_signal</code>). This can be another visualization for
helping us deciding the proper thresholds.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nCount_ATAC"</span>, <span class="st">"TSS.enrichment"</span>, <span class="st">"nucleosome_signal"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p><br> Now, based on our QC above, we can decide the thresholds for
filtering the low quality cells. Here, we will keep cells with number of
ATAC peak counts between 500 and 100k, number of RNA read counts between
500 and 12k, the strength of the nucleosome signal smaller than 1.5, and
the TTS enrichment score greater than 1.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter out low quality cells</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">50000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">12000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nucleosome_signal</span> <span class="op">&lt;</span> <span class="fl">1.5</span> <span class="op">&amp;</span></span>
<span>    <span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 193208 features across 2566 samples within 2 assays </span></span>
<span><span class="co">#&gt; Active assay: ATAC (156607 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  1 other assay present: RNA</span></span></code></pre></div>
<br><details><summary><strong><em>Question: How will you decide the thresholds for
cut-offs?</em></strong>
</summary><p><br> It is important to keep in mind, as with scRNA-seq we
introduced in our previous tutorial (Introduction to Single-cell), all
these cut-offs vary depending on your biological system, cell viability,
and other factors. In some cases, you may want to set lower or more
stringent thresholds to avoid removing rare cell populations or to
capture cell types which may have specific characteristics.</p>
In our previous tutorial (Introduction to Single-cell), we have walked
through the detailed QC procedure for scRNA-seq data. <em>Can you still
recall what they are?</em> In the current tutorial section, we
introduced another two scATAC-seq based QC metrics, which can help us
filtering the low quality cells from the perspective of the DNA
accessibility data. More scATAC-seq QC metrics can be found from the
Computing QC Metrics section in <a href="https://stuartlab.org/signac/articles/pbmc_vignette.html#computing-qc-metrics" class="external-link">Signac</a>.
</details>
</div>
<div class="section level3">
<h3 id="gene-expression-data-processing">Gene expression data processing<a class="anchor" aria-label="anchor" href="#gene-expression-data-processing"></a>
</h3>
<p>In our previous tutorial (Basics of Single-celll Annotation, Section
4), we’ve learned to preprocess the scRNA-seq data before dimension
reduction, using the <code>NormalizeData</code>,
<code>FindVariableFeatures</code> and <code>ScaleData</code> functions
from Seurat. More recently, Seurat proposes another normalization and
variance stabilization algorithm for preprocessing scRNA-seq data,
namely <a href="https://satijalab.org/seurat/reference/sctransform" class="external-link">SCTransform</a>,
which is recommended by Seurat to be used as the alternative to the
<code>NormalizeData</code>, <code>FindVariableFeatures</code>,
<code>ScaleData</code> workflow. Let’s try to use
<code>SCTransform</code> here. The results of <code>SCTransform</code>
are saved in a new assay (named <code>SCT</code> by default) with
<code>scale.data</code> being pearson residuals, which are claimed to be
independent of sequencing depth and can be used for different processing
and downstream analysis such as variable gene selection, dimensional
reduction, clustering, visualization, and differential expression. More
details can be found in their <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1" class="external-link">manuscript</a>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="co"># Let's try SCTransform as an alternative to the NormalizeData, FindVariableFeatures, ScaleData workflow</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co"># After that, we can run the PCA as we did before</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dna-accessibility-data-processing">DNA accessibility data processing<a class="anchor" aria-label="anchor" href="#dna-accessibility-data-processing"></a>
</h3>
<p>Now, we will process the DNA accessibility assay using a similar
workflow from Signac. We first apply the frequency-inverse document
frequency (TF-IDF) normalization (<a href="https://stuartlab.org/signac/reference/runtfidf" class="external-link">RunTFIDF</a>
function), which is a two-step normalization procedure that both
normalizes across cells to correct for differences in cellular
sequencing depth, and across peaks to give higher values to more rare
peaks. Then we find the set of variable features (<a href="https://stuartlab.org/signac/reference/findtopfeatures" class="external-link">FindTopFeatures</a>
function) and conduct the singular value decomposition (SVD) for
dimension reduction (<a href="https://stuartlab.org/signac/reference/runsvd" class="external-link">RunSVD</a>
function). The resulted dimension reduction object will be stored as
<code>lsi</code>. To put it simple, you can think of this as analogous
to the output of PCA in the scRNA-seq case.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="co"># TF-IDF normalization</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>, min.cutoff <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="co"># selecting the top features (with the minimum number of counts for the feature to be included set to be 5)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="co"># dimension reduction</span></span>
<span></span>
<span><span class="co"># RunTFIDF + RunSVD above performs the Latent Semantic Indexing (LSI), which is a dimension reduction method that uses term frequency inverse document frequency transformation (TFIDF) followed by singular value decomposition (SVD).</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h3>
<p>Try to recall that we have used the scClassify package to annotate
scRNA-seq datasets in our previous tutorial (Basics of Single-cell
Annotation). It is also possible to annotate our RNA+ATAC multiomic
dataset with scClassify, utilizing a reference scRNA-seq data with cell
type annotation, solely based on the transcriptome. More specifically,
we can train our scClassify model with our annotated reference scRNA-seq
data and predict the cell types of our <code>pbmc</code> dataset. This
is useful especially when we want to annotate our multiomic dataset but
only have a scRNA-seq atlas available at hand.</p>
<p>To do this, we will introduce another PBMC dataset generated from <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets/1.0.0/pbmc_granulocyte_sorted_10k" class="external-link">10x
Genomics</a>. Like our <code>pbmc</code> dataset, it is also a paired
multiomic data with scRNA-seq + scATAC-seq available. We annotated this
dataset based on the clusters, following the <a href="https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac" class="external-link">Seurat
Vignettes</a>. We can load the annotated reference data directly from
the <code>pbmc10k.rds</code> file as a Seurat object. We can see that
this reference data has been already normalized using the <a href="https://satijalab.org/seurat/reference/sctransform" class="external-link">SCTransform</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load PBMC reference</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'../data/pbmc10k.rds'</span><span class="op">)</span></span>
<span><span class="va">reference</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 166853 features across 10412 samples within 3 assays </span></span>
<span><span class="co">#&gt; Active assay: ATAC (106056 features, 106056 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  2 other assays present: RNA, SCT</span></span>
<span><span class="co">#&gt;  5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, wnn.umap</span></span></code></pre></div>
<p>We can access the annotation via <code>reference$celltype</code>.
From the bar plot below, we can see that it has 19 cell types.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">reference</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">celltype</span>, fill<span class="op">=</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>Now, let’s utilize the scRNA-seq from this reference pbmc dataset and
annotate our <code>pbmc</code> dataset using scClassify.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scClassify</span><span class="op">)</span></span>
<span><span class="co"># Train on reference pbmc dataset</span></span>
<span><span class="va">train_scClassify</span> <span class="op">&lt;-</span> <span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/train_scClassify.html" class="external-link">train_scClassify</a></span><span class="op">(</span>exprsMat_train <span class="op">=</span> <span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">$</span><span class="va">scale.data</span>, </span>
<span>                                           cellTypes_train <span class="op">=</span> <span class="va">reference</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>                                           algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                                           selectFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                                           similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>                                           returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                           verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; after filtering not expressed genes </span></span>
<span><span class="co">#&gt; [1]  3000 10412</span></span>
<span><span class="co">#&gt; [1] "Feature Selection..."</span></span>
<span><span class="co">#&gt; [1] "Number of genes selected to construct HOPACH tree 697"</span></span>
<span><span class="co">#&gt; [1] "Constructing tree ..."</span></span>
<span><span class="co">#&gt; [1] "Training...."</span></span>
<span><span class="co">#&gt; [1] "=== selecting features by: limma ===="</span></span></code></pre></div>
<p>After the training is finished, we can visualise the cell type
hierarchy tree.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/plotCellTypeTree.html" class="external-link">plotCellTypeTree</a></span><span class="op">(</span><span class="va">train_scClassify</span><span class="op">@</span><span class="va">cellTypeTree</span>, group_level <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>Then, we can predict the cell types of our <code>pbmc</code> dataset
using the trained scClassify model.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_res</span> <span class="op">&lt;-</span> <span class="fu">scClassify</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/predict_scClassify.html" class="external-link">predict_scClassify</a></span><span class="op">(</span>exprsMat_test <span class="op">=</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">$</span><span class="va">scale.data</span>,</span>
<span>                               trainRes <span class="op">=</span> <span class="va">train_scClassify</span>,</span>
<span>                               algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                               similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>                               prob_threshold <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Ensemble learning is disabled... </span></span>
<span><span class="co">#&gt; Using parameters: </span></span>
<span><span class="co">#&gt; similarity  algorithm   features </span></span>
<span><span class="co">#&gt;  "pearson"     "WKNN"    "limma" </span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; [1] "Using dynamic correlation cutoff..."</span></span>
<span><span class="co">#&gt; weights for each base method: </span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>Let’s look into the predicted cell types. We will save this
annotation into the <code>pbmc</code> object as
<code>annotation_scclassify</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pred_res</span><span class="op">$</span><span class="va">pearson_WKNN_limma</span><span class="op">$</span><span class="va">predLabelMat</span><span class="op">)</span></span>
<span><span class="va">annotation_scclassify</span> <span class="op">&lt;-</span> <span class="va">pred_res</span><span class="op">$</span><span class="va">pearson_WKNN_limma</span><span class="op">$</span><span class="va">predLabelMat</span><span class="op">[</span>,<span class="va">n</span><span class="op">]</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">annotation_scclassify</span> <span class="op">&lt;-</span> <span class="va">annotation_scclassify</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>celltype<span class="op">=</span><span class="va">annotation_scclassify</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">celltype</span>, fill<span class="op">=</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<p>Now, similar to what we’ve done with the scRNA-seq in our previous
tutorials, let’s try to run UMAP and visualize it with the cell type
annotation we just derived. Since we have two modalities, we can run
UMAP with both modalities.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction.name <span class="op">=</span> <span class="st">"umap.rna"</span>, reduction.key <span class="op">=</span> <span class="st">"rnaUMAP_"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction<span class="op">=</span><span class="st">'umap.rna'</span>, group.by <span class="op">=</span> <span class="st">"annotation_scclassify"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"lsi"</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span>, reduction.name <span class="op">=</span> <span class="st">"umap.atac"</span>, reduction.key <span class="op">=</span> <span class="st">"atacUMAP_"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction<span class="op">=</span><span class="st">'umap.atac'</span>, group.by <span class="op">=</span> <span class="st">"annotation_scclassify"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"ATAC"</span><span class="op">)</span>                  </span>
<span></span>
<span><span class="va">p1</span><span class="op">+</span><span class="va">p2</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="linking-peaks-to-genes">Linking peaks to genes<a class="anchor" aria-label="anchor" href="#linking-peaks-to-genes"></a>
</h3>
<p>For each gene, we can find the set of peaks that may regulate the
gene by by computing the correlation between gene expression and
accessibility at nearby peaks, and correcting for bias due to GC
content, overall accessibility, and peak size. See the <a href="https://www.biorxiv.org/content/10.1101/2020.11.09.373613v1" class="external-link">Signac
paper</a> for a full description of the method we use to link peaks to
genes.</p>
<p>Running this step on the whole genome can be time consuming, so here
we demonstrate peak-gene links for a subset of genes as an example. The
same function can be used to find links for all genes by omitting the
<code>genes.use</code> parameter:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">annotation_scclassify</span></span>
<span><span class="co"># first compute the GC content for each peak</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RegionStats.html" class="external-link">RegionStats</a></span><span class="op">(</span><span class="va">pbmc</span>, genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># link peaks to genes</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/LinkPeaks.html" class="external-link">LinkPeaks</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  peak.assay <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>  expression.assay <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  genes.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LYZ"</span>, <span class="st">"MS4A1"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can visualize these links using the <code><a href="https://stuartlab.org/signac/reference/CoveragePlot.html" class="external-link">CoveragePlot()</a></code>
function, or alternatively we could use the
<code><a href="https://stuartlab.org/signac/reference/CoverageBrowser.html" class="external-link">CoverageBrowser()</a></code> function in an interactive analysis:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idents.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Naive B"</span>, <span class="st">"Intermediate B"</span>, <span class="st">"Memory B"</span>,</span>
<span>                 <span class="st">"CD14 Mono"</span>, <span class="st">"CD16 Mono"</span>, <span class="st">"CD8 TEM_1"</span>, <span class="st">"CD8 TEM_2"</span>, <span class="st">"CD8 Naive"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CoveragePlot.html" class="external-link">CoveragePlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"MS4A1"</span>,</span>
<span>  features <span class="op">=</span> <span class="st">"MS4A1"</span>,</span>
<span>  expression.assay <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  idents <span class="op">=</span> <span class="va">idents.plot</span>,</span>
<span>  extend.upstream <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  extend.downstream <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CoveragePlot.html" class="external-link">CoveragePlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"LYZ"</span>,</span>
<span>  features <span class="op">=</span> <span class="st">"LYZ"</span>,</span>
<span>  expression.assay <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  idents <span class="op">=</span> <span class="va">idents.plot</span>,</span>
<span>  extend.upstream <span class="op">=</span> <span class="fl">8000</span>,</span>
<span>  extend.downstream <span class="op">=</span> <span class="fl">5000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save the pbmc data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">'../data/pbmc3k.rds'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="integrative-analysis-of-single-cell-multiomic-data-with-matilda">Integrative analysis of single-cell multiomic data with Matilda<a class="anchor" aria-label="anchor" href="#integrative-analysis-of-single-cell-multiomic-data-with-matilda"></a>
</h2>
<p>While methods designed for analysing scRNA-seq data or scATAC-seq
data alone can be also applied to analyse RNA modality or ATAC modality
in multimodal single-cell omics data, most of them cannot take advantage
of other available data modalities and therefore could not fully utilize
the information embedded in such data. This motivates the methods for
integrative analysis on multiomic data, which make the best use of all
these information across different modalities together in the downstream
tasks.</p>
<p>In this section, we will introduce a deep learning-based model called
<a href="https://watermark.silverchair.com/gkad157.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAA0kwggNFBgkqhkiG9w0BBwagggM2MIIDMgIBADCCAysGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMik7uvO7FTqABDeMQAgEQgIIC_CftO2hSzWxw6Vf7JkEAcSpYQwtOkIPxAQHA2nbU8zHvSzj8xsjbiOZuLg6psUgq6UCpDESP0zBy2XKBanSFg-kNVxuhpSkrbCVi2A7ncr3JGYxsoW_6ZJWQ-A7fcQ6zO4qgaUXeVgaSsY3bKDXbxqGc8_BnJgLac9Em9ysK9OizwjCI-K5irmBaisiyZJyLzdyTyfO9vv6YRPhybL29fCknvngzDPYek-tvj5X0URGQqqFFZPz5SQAyFjEt_fleCxOVBSR30kW7QbE4cv5pniI2iu8znCNUwo4jL-axYShy81amjf8TbjkrMQGDdtPH0z2NxmON39Vbhd-UjojNqCL77ZJf6k7LBgGvyRRBpkUfWOkpNTsnbtrFrpZ5aCFoNmaeoqesAlF15O-NSd-6rZhHANbx8-WFmusghj2FZrQdn_cCjt8qeGlwpx9VkV6gxelL3Mt4ftBCkAReq6aMVxJf1LypfcjO-QSoTJpe44bUC-bRV95vg7jToFLZR52K8a-JD8fZXtpDvn6mgN7nbzuD3gaqjgsWMlTUbXpVUpSUUz84LZx2FhXOFPUkNxXFjl0q9S9u9AlFSY1Mfwc46DlPwgYeRqi-7heHnA0iPMKIs_io4lOU9pcObDBixxSzk5OKaJh3GL0IWC0rzBNDrqMuILO5tAUCw2prUFZtXouc8hc_9mXZ906hi2BfMrAflzHxnaP2n-zA6RB1tfVJ2ZdnbR0Hzup4Hj6BHaOmZ8s9NqjbK6HsIjJ28PAJ16QNcT6-nCGbOULGFv1pWKYlWIMJORWDnzpV9-FGQtKeSyXvbjPwp1-5ELckUIgMy3Idkqc0KSQIZS789_rWvh-U0YPINFnxfY9AdWerbOaIHDXxMLjr88tMUHug0RmBT5GsrNYzy5rqqFJjXxIheCaNIqlNoywO5nTxWS2X9HQX_IeJC6LgD3q9uR_kFtiFCnQ-fvLSaO_HzQFpzZID3sDdFaYheTtU3RBCaCwZ1WxvM24L3Rlqj_TUzjh9qJRY" class="external-link">Matilda</a>,
which proposes a multitask learning method for integrative analysis of
multimodal single-cell omics data. It can be applied to integratively
analyse different combination of modalities such as RNA+ADT, RNA+ATAC
and RNA+ADT+ATAC etc. Thus, we can easily utilise it to analyse our 10x
multiome data with RNA+ATAC.</p>
<p>More specifically, we will use our reference RNA+ATAC data introduced
above as training and evaluation set while using our <code>pbmc</code>
dataset as test set to run Matilda. By doing this, we can utilise the
annotation from the reference dataset to annotate our <code>pbmc</code>
dataset, via looking at both modalities together, which is different to
our scClassify annotation in the previous section. Besides, we can also
conduct other integrative analysis on our <code>pbmc</code> dataset,
such as dimension reduction and clustering, as well as feature
selection.</p>
<p>Since Matilda is implemented in Python, we will move to our <a href="https://colab.research.google.com/drive/1aSU1Oi0ecBthcG27FvvKodxBg22B7Alo?usp=sharing" class="external-link">Python
environment for Matilda</a> for running the model. As is demonstrated in
Figure 11 below, we will first prepare our training and test data in
<code>R</code> here based on the input requirement of Matilda. Then we
will move to our <a href="https://colab.research.google.com/drive/1aSU1Oi0ecBthcG27FvvKodxBg22B7Alo?usp=sharing" class="external-link">Python
environment for Matilda</a> to train Matilda and run the our analytical
tasks. With the resulted outputs from running the analytical tasks, we
can then easily use our <code>R</code> pipeline to do the integrative
analysis with visualization.</p>
<p><strong>Note that, in order to save your time, we already prepared
the input data using the code provided in the follow-up sections and
made it accessible directly in our Python environment. Also, we already
trained Matilda and ran the analytical tasks and saved the outputs file
in our R environment here so that we can run the analysis code in the
follow-up sections directly. Thus, you don’t have to worry about the
data transfer between our R here and our Python environment at
all.</strong></p>
<div class="figure" style="text-align: center">
<img src="figures/matilda_workflow.png" alt="Figure 11. Matilda workflow" width="70%"><p class="caption">
Figure 11. Matilda workflow
</p>
</div>
<p>Now, let’s wrap our data into the input formats that are compatible
with Matilda. First, let us load the packages required for wrapping the
data for Matilda. The <code>CreateGeneActivityMatrix2.R</code> file
contains the functions for converting the ATAC peak counts into gene
activity score, which is recommended by Matilda. The
<code>data_to_h5.R</code> file contains the utility functions to write
the RNA/ATAC matrix and our cell type annotation into the required
formats.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"../utils/CreateGeneActivityMatrix2.R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"../utils/data_to_h5.R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span> <span class="co"># help us split the data into training and test</span></span></code></pre></div>
<p>Here is all the data we need to prepare in our case to run Matilda,
including the RNA and ATAC count matrix, and the cell type labels. When
we have our own multiomic datasets to be analysed, we can just prepare
like this. Easy!</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># prepare rna and peak count matrix</span></span>
<span><span class="va">rna_train_eval</span> <span class="op">&lt;-</span> <span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">peak_train_eval</span> <span class="op">&lt;-</span> <span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">$</span><span class="va">counts</span></span>
<span></span>
<span><span class="va">rna_test</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">peak_test</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">$</span><span class="va">counts</span></span>
<span></span>
<span><span class="co"># prepare the cell type annotation</span></span>
<span><span class="va">celltype_anno</span> <span class="op">&lt;-</span> <span class="va">reference</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[[</span><span class="st">'celltype'</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># names for saving the data</span></span>
<span><span class="va">data_name</span> <span class="op">&lt;-</span> <span class="st">'pbmc'</span></span>
<span><span class="va">anno_name</span> <span class="op">&lt;-</span> <span class="st">'seurat_annotation'</span></span></code></pre></div>
<p>After we have the data prepared, we can just simply run the following
code snippet, which</p>
<ul>
<li><p>converts the ATAC peak counts into gene activity score</p></li>
<li><p>filters the features (genes) as recommended by Matilda and select
the common features</p></li>
<li><p>splits (the reference) data into training and evaluation set (20%
and 80%) as is done in Maltilda manuscript (you can decide your own
ratio)</p></li>
<li><p>saves the data into required formats (<code>h5</code> for data
matrix and <code>csv</code> for cell type labels)</p></li>
</ul>
<p>Again, to save time, we already ran the code and prepared the data in
our <a href="https://colab.research.google.com/drive/1aSU1Oi0ecBthcG27FvvKodxBg22B7Alo?usp=sharing" class="external-link">Python
environment for Matilda</a>. We can easily download the prepared data in
there based on the instructions.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Convert peak count to gene activity score</span></span>
<span><span class="co"># We can easily download other versions of the Ensembl human gene annotations and pass its path as the 'annotation.file' parameter</span></span>
<span><span class="va">gene.activities_train_eval</span> <span class="op">&lt;-</span> <span class="fu">CreateGeneActivityMatrix2</span><span class="op">(</span>peak.matrix<span class="op">=</span><span class="va">peak_train_eval</span>,</span>
<span>                                             annotation.file <span class="op">=</span><span class="st">"../data/Homo_sapiens.GRCh37.82.gtf"</span>,</span>
<span>                                             seq.levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">19</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span>,seq_replace<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene.activities_test</span> <span class="op">&lt;-</span> <span class="fu">CreateGeneActivityMatrix2</span><span class="op">(</span>peak.matrix<span class="op">=</span><span class="va">peak_test</span>,</span>
<span>                                             annotation.file <span class="op">=</span><span class="st">"../data/Homo_sapiens.GRCh37.82.gtf"</span>,</span>
<span>                                             seq.levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">19</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span>,seq_replace<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_train_eval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_train_eval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_train_eval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_train_eval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_test</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_test</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filtered out RNA and ATAC quantified in fewer than 1% of the cells</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">rna_train_eval</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rna_train_eval</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rna_train_eval</span> <span class="op">&lt;-</span> <span class="va">rna_train_eval</span><span class="op">[</span><span class="va">sel</span>,<span class="op">]</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">gene.activities_train_eval</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gene.activities_train_eval</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene.activities_train_eval</span> <span class="op">&lt;-</span> <span class="va">gene.activities_train_eval</span><span class="op">[</span><span class="va">sel</span>,<span class="op">]</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">rna_test</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rna_test</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rna_test</span> <span class="op">&lt;-</span> <span class="va">rna_test</span><span class="op">[</span><span class="va">sel</span>,<span class="op">]</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">gene.activities_test</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gene.activities_test</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene.activities_test</span> <span class="op">&lt;-</span> <span class="va">gene.activities_test</span><span class="op">[</span><span class="va">sel</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Keep the common genes/features</span></span>
<span><span class="va">common_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_train_eval</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_train_eval</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rna_test</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.activities_test</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rna_train_eval</span> <span class="op">&lt;-</span> <span class="va">rna_train_eval</span><span class="op">[</span><span class="va">common_gene</span>,<span class="op">]</span></span>
<span><span class="va">gene.activities_train_eval</span> <span class="op">&lt;-</span> <span class="va">gene.activities_train_eval</span><span class="op">[</span><span class="va">common_gene</span>, <span class="op">]</span></span>
<span><span class="va">rna_test</span> <span class="op">&lt;-</span> <span class="va">rna_test</span><span class="op">[</span><span class="va">common_gene</span>,<span class="op">]</span></span>
<span><span class="va">gene.activities_test</span> <span class="op">&lt;-</span> <span class="va">gene.activities_test</span><span class="op">[</span><span class="va">common_gene</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># filter out cell types that have less than 10 cells</span></span>
<span><span class="va">rna_train_eval.filt</span> <span class="op">&lt;-</span> <span class="va">rna_train_eval</span><span class="op">[</span>, <span class="va">celltype_anno</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">gene.activities_train_eval.filt</span> <span class="op">&lt;-</span> <span class="va">gene.activities_train_eval</span><span class="op">[</span>, <span class="va">celltype_anno</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">celltype_anno.filt</span> <span class="op">&lt;-</span> <span class="va">celltype_anno</span><span class="op">[</span><span class="va">celltype_anno</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celltype_anno</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Split the data to get our training and test data (20%/80%)</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html" class="external-link">createFolds</a></span><span class="op">(</span><span class="va">celltype_anno.filt</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">folds.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">f</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># 20%</span></span>
<span><span class="va">folds.eval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">f</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">f</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, <span class="va">f</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>, <span class="va">f</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># 80%</span></span>
<span></span>
<span><span class="va">rna.train</span> <span class="op">&lt;-</span> <span class="va">rna_train_eval.filt</span><span class="op">[</span>, <span class="va">folds.train</span><span class="op">]</span></span>
<span><span class="va">rna.eval</span> <span class="op">&lt;-</span> <span class="va">rna_train_eval.filt</span><span class="op">[</span>, <span class="va">folds.eval</span><span class="op">]</span></span>
<span><span class="va">gene.activities.train</span> <span class="op">&lt;-</span> <span class="va">gene.activities_train_eval.filt</span><span class="op">[</span>, <span class="va">folds.train</span><span class="op">]</span></span>
<span><span class="va">gene.activities.eval</span> <span class="op">&lt;-</span> <span class="va">gene.activities_train_eval.filt</span><span class="op">[</span>, <span class="va">folds.eval</span><span class="op">]</span></span>
<span><span class="va">celltype_anno.train</span> <span class="op">&lt;-</span> <span class="va">celltype_anno.filt</span><span class="op">[</span><span class="va">folds.train</span><span class="op">]</span></span>
<span><span class="va">celltype_anno.eval</span> <span class="op">&lt;-</span> <span class="va">celltype_anno.filt</span><span class="op">[</span><span class="va">folds.eval</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># save processed data</span></span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna.train</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/rna_"</span>,<span class="va">data_name</span>,<span class="st">"_train.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna.eval</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/rna_"</span>,<span class="va">data_name</span>,<span class="st">"_eval.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna_test</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/rna_"</span>,<span class="va">data_name</span>,<span class="st">"_test.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>atac <span class="op">=</span> <span class="va">gene.activities.train</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/gas_"</span>,<span class="va">data_name</span>,<span class="st">"_train.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>atac <span class="op">=</span> <span class="va">gene.activities.eval</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/gas_"</span>,<span class="va">data_name</span>,<span class="st">"_eval.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">write_h5</span><span class="op">(</span>exprs_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>atac <span class="op">=</span> <span class="va">gene.activities_test</span><span class="op">)</span>, </span>
<span>             h5file_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/gas_"</span>,<span class="va">data_name</span>,<span class="st">"_test.h5"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save annotation</span></span>
<span><span class="fu">write_csv</span><span class="op">(</span>cellType_list <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cty <span class="op">=</span> <span class="va">celltype_anno.train</span><span class="op">)</span>,</span>
<span>            csv_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/"</span>,<span class="va">data_name</span>,<span class="st">"_"</span>,<span class="va">anno_name</span>, <span class="st">"_cty_train.csv"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu">write_csv</span><span class="op">(</span>cellType_list <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cty <span class="op">=</span> <span class="va">celltype_anno.eval</span><span class="op">)</span>,</span>
<span>            csv_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"../data/"</span>,<span class="va">data_name</span>,<span class="st">"_"</span>,<span class="va">anno_name</span>, <span class="st">"_cty_eval.csv"</span>,sep<span class="op">=</span><span class="st">''</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<div class="section level3">
<h3 id="play-with-matilda">Play with Matilda<a class="anchor" aria-label="anchor" href="#play-with-matilda"></a>
</h3>
<p>Let’s go to <a href="https://colab.research.google.com/drive/1aSU1Oi0ecBthcG27FvvKodxBg22B7Alo?usp=sharing" class="external-link">Python
environment for Matilda</a>.</p>
</div>
<div class="section level3">
<h3 id="integrative-analysis-with-visualization">Integrative analysis with visualization<a class="anchor" aria-label="anchor" href="#integrative-analysis-with-visualization"></a>
</h3>
<p>Welcome back!</p>
<p>Now, we’ve finished running Matilda with analytical tasks. All of our
Matilda output files (i.e., results generated from our Python
environment) are saved in the <code>data/output/</code> folder. We can
analyse these results with visualizations easily. Let’s first import the
libraries.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/" class="external-link">mclust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jchiquet/aricode" class="external-link">aricode</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tkonopka/umap" class="external-link">umap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sthda.com/english/wiki/ggcorrplot-visualization-of-a-correlation-matrix-using-ggplot2" class="external-link">ggcorrplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/rhdf5" class="external-link">rhdf5</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="data-simulation">Data simulation<a class="anchor" aria-label="anchor" href="#data-simulation"></a>
</h4>
<p>Try to recall that, in our Python environment, we’ve run the data
simulation on our training data for both (1) all cell types, and (2) the
3 specific cell types, including CD16 Mono, CD8 Naive and NK. Following
the <a href="https://watermark.silverchair.com/gkad157.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAA0kwggNFBgkqhkiG9w0BBwagggM2MIIDMgIBADCCAysGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMpBIHNKaEj1hcHlLqAgEQgIIC_DyOVLWNHW5FQza5YiSdpxYtJEbo8Y6PlCubErGTdfBd5wpolSlM-dXuQ0PNv_VNoYR26DFpx51j61np908Avn6U7CVaqWyeU_Kza5029BRDugWc_FKmKc-r2TUTnvi8cFOnPgDbfbjBxp77VYl8JrkIxX5G7JSsSh6CeG0PYHchYBFHVpy9f-6Ox4V24kBmuPiIHLmMM2AS9m1cNeBG8FVwQ-BDvEkg8wsq7RPJTz8nuWpP6OHlfeUcFjAfg6x2SG566yDoCdjOxX1odLrwVuZs1CVcER7_ZHsLGjKSXbmEbPes-8yTYdmuhIR1rAzjoxeLjQObSpeTYm5J7f3ilhCwcgg5nQY6lNaLCL7zV7Z5TGHT7oOwupC6zr-tlGMbKK9ocKR0CqJN7KMyvRy1_jgTfVwQrho_e2yjZCJDxhuRlIX45TTlXo-Dy4j6KMl9a_C1OCB7mc48ZDkECO0YoPHgIUQEw2RI2JCb2wIrirksOoL5etqgpqlqu-hj1YPxZw3LoNeE1XEzbp_OzmC0IIpJtGbrui9KO0hGyAzd9AQ68s-d-nV3xHb69oz5tgdw-NAnPveOI_W-He6gyGHQ0YbkGnSy_CRgPvF_mlk88x4maKKyjR5ByqX8Im6OcOg3Tg_lWzAi2TJVRIyhfRMLGiCaZwZjDhDDHs1rlZuM8RHuF-5gsi25cBzNptuoXdvg25oNeCbkhlPd96ptdBuDTS3ekwdsS0icLR2ctJykHQCy3WiERvt5M-HlXNLvltFJm-p1jB6pphMGBc6O-pGMlJr3EXMhrwK5MRK_3wl7mtKgjuyIPBuUA236lbEdDkYX01sgCfhngdo5S8LXCXyrtH6-qrGbOA8mxljr1UcIWTaVkbRXAt6bxynY3RzvPk-yla6Rrb-thf1vWUz32NDT4I7QRCVzMAkavcnVf-qBywGvWmdNy98iaWWqu1IylQc_TCWTSd3k7XQmN66BBIN9hnh7vHnYF7Diaf8NLEWkCLq4PpBDu8h2ebMqsV33#page=1.11" class="external-link">manuscript</a>,
we can inspect the quality of the simulated data based on its Figure 2
A-C and Figure 1 B.</p>
<p>For inspecting the simulation for all the cell types, we can plot the
heatmaps which visualize the correlation structure of selected features
in each data modality (See Figure 2 A-C in the <a href="https://watermark.silverchair.com/gkad157.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAA0kwggNFBgkqhkiG9w0BBwagggM2MIIDMgIBADCCAysGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMpBIHNKaEj1hcHlLqAgEQgIIC_DyOVLWNHW5FQza5YiSdpxYtJEbo8Y6PlCubErGTdfBd5wpolSlM-dXuQ0PNv_VNoYR26DFpx51j61np908Avn6U7CVaqWyeU_Kza5029BRDugWc_FKmKc-r2TUTnvi8cFOnPgDbfbjBxp77VYl8JrkIxX5G7JSsSh6CeG0PYHchYBFHVpy9f-6Ox4V24kBmuPiIHLmMM2AS9m1cNeBG8FVwQ-BDvEkg8wsq7RPJTz8nuWpP6OHlfeUcFjAfg6x2SG566yDoCdjOxX1odLrwVuZs1CVcER7_ZHsLGjKSXbmEbPes-8yTYdmuhIR1rAzjoxeLjQObSpeTYm5J7f3ilhCwcgg5nQY6lNaLCL7zV7Z5TGHT7oOwupC6zr-tlGMbKK9ocKR0CqJN7KMyvRy1_jgTfVwQrho_e2yjZCJDxhuRlIX45TTlXo-Dy4j6KMl9a_C1OCB7mc48ZDkECO0YoPHgIUQEw2RI2JCb2wIrirksOoL5etqgpqlqu-hj1YPxZw3LoNeE1XEzbp_OzmC0IIpJtGbrui9KO0hGyAzd9AQ68s-d-nV3xHb69oz5tgdw-NAnPveOI_W-He6gyGHQ0YbkGnSy_CRgPvF_mlk88x4maKKyjR5ByqX8Im6OcOg3Tg_lWzAi2TJVRIyhfRMLGiCaZwZjDhDDHs1rlZuM8RHuF-5gsi25cBzNptuoXdvg25oNeCbkhlPd96ptdBuDTS3ekwdsS0icLR2ctJykHQCy3WiERvt5M-HlXNLvltFJm-p1jB6pphMGBc6O-pGMlJr3EXMhrwK5MRK_3wl7mtKgjuyIPBuUA236lbEdDkYX01sgCfhngdo5S8LXCXyrtH6-qrGbOA8mxljr1UcIWTaVkbRXAt6bxynY3RzvPk-yla6Rrb-thf1vWUz32NDT4I7QRCVzMAkavcnVf-qBywGvWmdNy98iaWWqu1IylQc_TCWTSd3k7XQmN66BBIN9hnh7vHnYF7Diaf8NLEWkCLq4PpBDu8h2ebMqsV33#page=1.11" class="external-link">manuscript</a>).
Specifically, we first apply the functions <code>modelGeneVar</code> and
<code>getTopHVGs</code> from the <code>scran</code> R package to select
the top N highly variable genes (HVGs) based on their variability
calculated from each data modality. Then we calculate pairwise Pearson’s
correlation coefficients from these HVGs across all cells.</p>
<p>We provide the following
<code>plot_simulation_cor_single_modality</code> function, which can
plot this heatmap for the specified modality (e.g., RNA or ATAC) based
on the Matilda output files. We only need to input the following
parameters:</p>
<ul>
<li>
<code>simulation_folder</code>: the data folder which contains the
output files from running the Matilda simulation task. In our case, it
will be the
<code>data/output/simulation_result/SHAREseq/all_cell_type/</code> for
the all cell type simulation.</li>
<li>
<code>simulated_cell_type</code>: the specific cell type to be
visualized. We will ignore this parameter since we’re inspecting the
simulation for all cell type.</li>
<li>
<code>modality</code>: the specific modality to be visualized. In
our case, it will be either <code>RNA</code> or <code>ATAC</code>.</li>
<li>
<code>topn_hvg</code>: the number of HVGs to be included in the
heatmap. We will use <code>50</code> for demonstration. You can adjust
by yourself.</li>
</ul>
<p>This function will return the two heatmap plots, one for the real
data and the other for the Matilda-simulated data, which can be easily
visualized together and compared.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot_simulation_cor_single_modality</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">simulation_folder</span>, <span class="va">simulated_cell_type</span><span class="op">=</span><span class="st">''</span>, <span class="va">modality</span><span class="op">=</span><span class="st">'RNA'</span>, <span class="va">topn_hvg</span><span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">real_label_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"real_label.csv"</span><span class="op">)</span></span>
<span>  <span class="va">real_data_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"real_data_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">modality</span><span class="op">)</span>,<span class="st">".csv"</span><span class="op">)</span></span>
<span>  <span class="va">sim_label_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"sim_label.csv"</span><span class="op">)</span></span>
<span>  <span class="va">sim_data_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"sim_data_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">modality</span><span class="op">)</span>,<span class="st">".csv"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">real_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">real_label_csv</span><span class="op">)</span></span>
<span>  <span class="va">real_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">real_data_csv</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">sim_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">sim_label_csv</span><span class="op">)</span></span>
<span>  <span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">sim_data_csv</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">var.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">real_data</span><span class="op">)</span></span>
<span>  <span class="va">hvgs_rna</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">var.out</span>, n <span class="op">=</span> <span class="va">topn_hvg</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">simulated_cell_type</span> <span class="op">==</span> <span class="st">''</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cor_train.sim.data</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">[</span><span class="va">hvgs_rna</span>,<span class="op">]</span></span>
<span>    <span class="va">cor_train.real.data</span> <span class="op">&lt;-</span> <span class="va">real_data</span><span class="op">[</span><span class="va">hvgs_rna</span>,<span class="op">]</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">cor_train.sim.data</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">[</span><span class="va">hvgs_rna</span>,<span class="va">sim_label</span><span class="op">$</span><span class="va">label</span><span class="op">==</span><span class="va">simulated_cell_type</span><span class="op">]</span></span>
<span>    <span class="va">cor_train.real.data</span> <span class="op">&lt;-</span> <span class="va">real_data</span><span class="op">[</span><span class="va">hvgs_rna</span>,<span class="va">real_label</span><span class="op">$</span><span class="va">label</span><span class="op">==</span><span class="va">simulated_cell_type</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">cor_sim_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cor_train.sim.data</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cor_train.sim.data</span><span class="op">)</span>, use<span class="op">=</span><span class="st">"na.or.complete"</span><span class="op">)</span></span>
<span>  <span class="va">cor_real_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cor_train.real.data</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cor_train.real.data</span><span class="op">)</span>, use<span class="op">=</span><span class="st">"na.or.complete"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">dd</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">cor_real_mat</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">dd</span>, method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span></span>
<span>  <span class="va">cor_order</span> <span class="op">&lt;-</span> <span class="va">hc</span><span class="op">$</span><span class="va">order</span></span>
<span></span>
<span>  <span class="va">cor_real_mat</span> <span class="op">&lt;-</span> <span class="va">cor_real_mat</span><span class="op">[</span><span class="va">cor_order</span>,<span class="va">cor_order</span><span class="op">]</span></span>
<span></span>
<span>  <span class="va">plot_cor1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html" class="external-link">ggcorrplot</a></span><span class="op">(</span><span class="va">cor_real_mat</span>,hc.order <span class="op">=</span> <span class="cn">FALSE</span>, hc.method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>limit<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span>,mid<span class="op">=</span><span class="st">"white"</span> <span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"real {modality} (top {topn_hvg} hvg)"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">cor_sim_mat</span> <span class="op">&lt;-</span> <span class="va">cor_sim_mat</span><span class="op">[</span><span class="va">cor_order</span>,<span class="va">cor_order</span><span class="op">]</span></span>
<span></span>
<span>  <span class="va">plot_cor2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html" class="external-link">ggcorrplot</a></span><span class="op">(</span><span class="va">cor_sim_mat</span>,hc.order <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"Matilda-simulated {modality} (top {topn_hvg} hvg)"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>limit<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span>,mid<span class="op">=</span><span class="st">"white"</span> <span class="op">)</span></span>
<span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">plot_cor1</span>,<span class="va">plot_cor2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>  </span></code></pre></div>
<p>Let’s run the <code>plot_simulation_cor_single_modality</code> for
both RNA and ATAC and plot the results. In the visualization below, the
left-upper plot and left-bottom plot are the correlation structure
heatmap for the real RNA data and Matilda-simulated RNA data,
respectively. Similarly, the right-upper plot and right-bottom plot are
the correlation structure heatmap for the real ATAC data and
Matilda-simulated ATAC data, respectively. As expected, we can clearly
see the similar pattern between the real data and Matilda-simulated
data.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### specify the data folder which contains the Matilda-simulated output to be used for visualization</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/simulation_result/SHAREseq/all_cell_type/"</span></span>
<span></span>
<span><span class="co">### plot heatmap for real and simulated RNA</span></span>
<span><span class="va">rna_cor_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_cor_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, modality<span class="op">=</span><span class="st">'RNA'</span>, topn_hvg<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">### plot heatmap for real and simulated ATAC</span></span>
<span><span class="va">atac_cor_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_cor_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, modality<span class="op">=</span><span class="st">'ATAC'</span>, topn_hvg<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### visualize the heatmaps together</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">rna_cor_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_cor_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">rna_cor_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_cor_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p>For inspecting the simulation for specific cell type, we can plot the
UMAP visualization to compare the real and Matilda simulated cells for
this anchor cell type (See Figure 1 B in the <a href="https://watermark.silverchair.com/gkad157.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAA0kwggNFBgkqhkiG9w0BBwagggM2MIIDMgIBADCCAysGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMpBIHNKaEj1hcHlLqAgEQgIIC_DyOVLWNHW5FQza5YiSdpxYtJEbo8Y6PlCubErGTdfBd5wpolSlM-dXuQ0PNv_VNoYR26DFpx51j61np908Avn6U7CVaqWyeU_Kza5029BRDugWc_FKmKc-r2TUTnvi8cFOnPgDbfbjBxp77VYl8JrkIxX5G7JSsSh6CeG0PYHchYBFHVpy9f-6Ox4V24kBmuPiIHLmMM2AS9m1cNeBG8FVwQ-BDvEkg8wsq7RPJTz8nuWpP6OHlfeUcFjAfg6x2SG566yDoCdjOxX1odLrwVuZs1CVcER7_ZHsLGjKSXbmEbPes-8yTYdmuhIR1rAzjoxeLjQObSpeTYm5J7f3ilhCwcgg5nQY6lNaLCL7zV7Z5TGHT7oOwupC6zr-tlGMbKK9ocKR0CqJN7KMyvRy1_jgTfVwQrho_e2yjZCJDxhuRlIX45TTlXo-Dy4j6KMl9a_C1OCB7mc48ZDkECO0YoPHgIUQEw2RI2JCb2wIrirksOoL5etqgpqlqu-hj1YPxZw3LoNeE1XEzbp_OzmC0IIpJtGbrui9KO0hGyAzd9AQ68s-d-nV3xHb69oz5tgdw-NAnPveOI_W-He6gyGHQ0YbkGnSy_CRgPvF_mlk88x4maKKyjR5ByqX8Im6OcOg3Tg_lWzAi2TJVRIyhfRMLGiCaZwZjDhDDHs1rlZuM8RHuF-5gsi25cBzNptuoXdvg25oNeCbkhlPd96ptdBuDTS3ekwdsS0icLR2ctJykHQCy3WiERvt5M-HlXNLvltFJm-p1jB6pphMGBc6O-pGMlJr3EXMhrwK5MRK_3wl7mtKgjuyIPBuUA236lbEdDkYX01sgCfhngdo5S8LXCXyrtH6-qrGbOA8mxljr1UcIWTaVkbRXAt6bxynY3RzvPk-yla6Rrb-thf1vWUz32NDT4I7QRCVzMAkavcnVf-qBywGvWmdNy98iaWWqu1IylQc_TCWTSd3k7XQmN66BBIN9hnh7vHnYF7Diaf8NLEWkCLq4PpBDu8h2ebMqsV33#page=1.11" class="external-link">manuscript</a>).</p>
<p>The following <code>plot_simulation_umap_single_modality</code> can
achieve this goal for us based on the Matilda simulation output files.
The input parameters are:</p>
<ul>
<li><p><code>simulation_folder</code>: the data folder which contains
the output files from running the Matilda simulation task. In our case,
it will be the
<code>data/output/simulation_result/SHAREseq/{cell type name}/</code>
for the the specific cell type simulation.</p></li>
<li><p><code>simulated_cell_type</code>: the specific cell type to be
visualized. In our case, it will be <code>CD16 Mono</code>,
<code>CD8 Naive</code> or <code>NK</code>.</p></li>
<li><p><code>modality</code>: the specific modality to be visualized. In
our case, it will be either <code>RNA</code> or
<code>ATAC</code>.</p></li>
</ul>
<p>This function will return the two UMAP plots, one for the real data
and the other for the Matilda-simulated data, which can be easily
visualized together and compared.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_simulation_umap_single_modality</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">simulation_folder</span>, <span class="va">simulated_cell_type</span>, <span class="va">modality</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">real_label_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"real_label.csv"</span><span class="op">)</span></span>
<span>  <span class="va">real_data_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"real_data_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">modality</span><span class="op">)</span>,<span class="st">".csv"</span><span class="op">)</span></span>
<span>  <span class="va">sim_label_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"sim_label.csv"</span><span class="op">)</span></span>
<span>  <span class="va">sim_data_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">simulation_folder</span>,<span class="st">"sim_data_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">modality</span><span class="op">)</span>,<span class="st">".csv"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">real_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">real_label_csv</span><span class="op">)</span></span>
<span>  <span class="va">train.real</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">real_data_csv</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"sim_{c(1:dim(train.real)[2])}"</span><span class="op">)</span></span>
<span>  <span class="va">sim_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">sim_label_csv</span><span class="op">)</span></span>
<span>  <span class="va">train.sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">sim_data_csv</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">train.sim</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"sim_{c(1:dim(train.sim)[2])}"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">train.cty.real</span> <span class="op">&lt;-</span> <span class="va">real_label</span><span class="op">$</span><span class="va">label</span></span>
<span>  <span class="va">train.cty.real</span><span class="op">[</span><span class="va">train.cty.real</span><span class="op">!=</span><span class="va">simulated_cell_type</span><span class="op">]</span> <span class="op">=</span> <span class="st">"other cell types"</span> </span>
<span>  <span class="va">train.cty.real</span><span class="op">[</span><span class="va">train.cty.real</span><span class="op">==</span><span class="va">simulated_cell_type</span><span class="op">]</span> <span class="op">=</span> <span class="st">"anchor cell type"</span></span>
<span></span>
<span>  <span class="va">train.cty.sim</span> <span class="op">&lt;-</span> <span class="va">sim_label</span><span class="op">$</span><span class="va">label</span></span>
<span>  <span class="va">train.cty.sim</span><span class="op">[</span><span class="va">train.cty.sim</span><span class="op">!=</span><span class="va">simulated_cell_type</span><span class="op">]</span> <span class="op">=</span> <span class="st">"other cell types"</span> </span>
<span>  <span class="va">train.cty.sim</span><span class="op">[</span><span class="va">train.cty.sim</span><span class="op">==</span><span class="va">simulated_cell_type</span><span class="op">]</span> <span class="op">=</span> <span class="st">"anchor cell type"</span> </span>
<span></span>
<span>  <span class="va">data.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">train.real</span>, <span class="va">train.sim</span><span class="op">)</span></span>
<span>  <span class="va">sce.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts<span class="op">=</span><span class="va">data.combined</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">sce.combined</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/ReadH5.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"original"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"augment"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.sim</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">sce.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce.combined</span>, n_threads<span class="op">=</span><span class="fl">10</span>, n_neighbors <span class="op">=</span> <span class="fl">20</span>, scale<span class="op">=</span><span class="cn">TRUE</span>, ntop<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">umap_both_layout</span> <span class="op">&lt;-</span> <span class="va">sce.combined</span><span class="op">@</span><span class="va">int_colData</span><span class="op">$</span><span class="va">reducedDims</span><span class="op">$</span><span class="va">UMAP</span></span>
<span>  <span class="va">umap_both_layout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">umap_both_layout</span><span class="op">)</span></span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"2"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"4"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.sim</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">umap_both_layout1</span> <span class="op">&lt;-</span> <span class="va">umap_both_layout</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="op">]</span></span>
<span>  <span class="va">umap_both_layout2</span> <span class="op">&lt;-</span> <span class="va">umap_both_layout</span><span class="op">[</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.real</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">train.sim</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span>  <span class="va">train.cty.sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/ReadH5.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">train.cty.sim</span><span class="op">)</span></span>
<span>  <span class="va">train.cty.real</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/ReadH5.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">train.cty.real</span><span class="op">)</span></span>
<span>  <span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_both_layout1</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">train.cty.real</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"Real {modality}"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_both_layout2</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">train.cty.sim</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"Matilda-simulated {modality}"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">plot1</span>,<span class="va">plot2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Let’s first try with CD16 Mono. In the visualization output, the
left-upper and left-bottom plots are the UMAPs for the real RNA data and
the RNA data with the cells of specific anchor cell type replaced with
corresponding Matilda-simulated cells. Similarly, right-upper and
right-bottom plots are the UMAPs for the real ATAC data and the ATAC
data with the cells of specific anchor cell type replaced with
corresponding Matilda-simulated cells. We can see that, for both RNA and
ATAC, the Matilda simulated cells are very similar to the real data, as
expected.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### specify the data folder which contains the Matilda-simulated output to be used for visualization</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/simulation_result/SHAREseq/CD16_Mono/"</span></span>
<span></span>
<span><span class="co">### specify the anchor cell type</span></span>
<span><span class="va">target_cell_type</span> <span class="op">&lt;-</span> <span class="st">'CD16 Mono'</span></span>
<span></span>
<span><span class="co">### plot UMAP for real and simulated RNA</span></span>
<span><span class="va">rna_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'RNA'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### plot UMAP for real and simulated ATAC</span></span>
<span><span class="va">atac_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'ATAC'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### visualize all plots together</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">rna_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">rna_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>Similarly, we can visualize for CD8 Naive.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/simulation_result/SHAREseq/CD8_Naive/"</span></span>
<span><span class="va">target_cell_type</span> <span class="op">&lt;-</span> <span class="st">'CD8 Naive'</span></span>
<span></span>
<span><span class="co">### plot real and simulated RNA</span></span>
<span><span class="va">rna_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'RNA'</span><span class="op">)</span></span>
<span><span class="co">### plot real and simulated ATAC</span></span>
<span><span class="va">atac_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'ATAC'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">rna_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">rna_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>Similarly, we can visualize for NK.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/simulation_result/SHAREseq/NK/"</span></span>
<span><span class="va">target_cell_type</span> <span class="op">&lt;-</span> <span class="st">'NK'</span></span>
<span></span>
<span><span class="co">### plot real and simulated RNA</span></span>
<span><span class="va">rna_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'RNA'</span><span class="op">)</span></span>
<span><span class="co">### plot real and simulated ATAC</span></span>
<span><span class="va">atac_plots</span> <span class="op">&lt;-</span> <span class="fu">plot_simulation_umap_single_modality</span><span class="op">(</span>simulation_folder<span class="op">=</span><span class="va">data_folder</span>, simulated_cell_type<span class="op">=</span><span class="va">target_cell_type</span>, modality<span class="op">=</span><span class="st">'ATAC'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">rna_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">rna_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">atac_plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="cell-type-annotation-1">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation-1"></a>
</h4>
<p>In our Python environment, we’ve run cell type classification on both
evaluation set (with ground truth label) and test set (without ground
truth label). Let’s first have a look at the evaluation set.</p>
<p>We can find the cell type classification output files for our
evaluation set from
<code>data/output/classification/SHAREseq/eval/</code>. In the folder,
we can see Matilda has generated the following two result files:</p>
<ul>
<li><p><code>accuracy_each_ct.txt</code>: This file saves the the
classification <code>Accuracy</code> for each cell type.</p></li>
<li><p><code>accuracy_each_cell.txt</code>: This file contains the
detailed key information regarding the classification results of each
individual cells, including the <code>real cell type</code> and
<code>predicted cell type</code>. We can easily parse the file to get
the ground truth label and Matilda-predicted label for
comparison.</p></li>
</ul>
<p>First, let’s have a look at the overall prediction performance
(accuracy) for each cell type. From the bar chart below, we can see from
the overall trend that the Plasma cells got 100% prediction accuracy
whereas the gdT cells (Gamma delta T cells) got the lowest accuracy of
around 37.39%.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the data folder that contains the classification output files from Matilda </span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/classification/SHAREseq/eval/"</span></span>
<span></span>
<span><span class="co"># Parse the accuracy_each_ct.txt file to get accuracy for each cell type</span></span>
<span><span class="va">acc_each_cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="st">"accuracy_each_ct.txt"</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">acc_each_cell</span>, sep<span class="op">=</span><span class="st">'\t'</span><span class="op">)</span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="st">": "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">5</span><span class="op">]</span>, <span class="st">", "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="st">'\\('</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Plot the bar chart for accuracy, ordered from high to low.</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>CellType <span class="op">=</span> <span class="va">cell_types</span>, Accuracy <span class="op">=</span> <span class="va">accuracy</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">CellType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">CellType</span>, levels <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">CellType</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">$</span><span class="va">Accuracy</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CellType</span>, y <span class="op">=</span> <span class="va">Accuracy</span>, fill <span class="op">=</span> <span class="va">CellType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Classification Accuracy for Each Cell Types"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Cell Type"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Accuracy"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="co"># Set the y-axis limits from 0 to 1 if accuracy is in this range</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Accuracy</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>Now, we can take a closer look at the predicted cell types VS. ground
truth cell types via a confusion matrix, which gives us more details
regarding the prediction performance/accuracy for each cell type. In the
confusion matrix below, the vertical labels represent the predicted cell
types and the horizontal labels represent the actual cell types. The
corresponding value in the matrix specifies the number of cells that are
predicted to be the specific cell type based on its row label while
being expected to be the specific cell type based on its column label. A
good classification performance expects the diagonal line to be mostly
highlighted, which can be seen clearly from the Matilda classification
output in the confusion matrix. By looking into the gdT cells (Gamma
delta T cells), we can see that most of the wrongly classified gdT cells
are predicted as CD8 TEM cells, which might be expected.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Specify the data folder that contains the classification output files from Matilda</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/classification/SHAREseq/eval/"</span></span>
<span></span>
<span><span class="co"># Parse the accuracy_each_cell.txt file to get the real label and predicted label for comparison</span></span>
<span><span class="va">acc_each_cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="st">"accuracy_each_cell.txt"</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">acc_each_cell</span>, sep<span class="op">=</span><span class="st">'\t'</span><span class="op">)</span></span>
<span><span class="va">real_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="st">": "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">predicted_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">5</span><span class="op">]</span>, <span class="st">": "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the confusion matrix based on the real label and predicted label</span></span>
<span><span class="va">conf_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/confusionMatrix.html" class="external-link">confusionMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">predicted_label</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">real_label</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">conf_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">conf_matrix</span><span class="op">$</span><span class="va">table</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">conf_table</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reference</span>, y <span class="op">=</span> <span class="va">Prediction</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">Freq</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Freq</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Confusion Matrix"</span>, x <span class="op">=</span> <span class="st">"Actual"</span>, y <span class="op">=</span> <span class="st">"Predicted"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
<p>Now, let’s move to the prediction results on our test set. Since we
don’t have ground truth labels to compare with the predicted results, we
will extract the predicted cell types only from the
<code>accuracy_each_cell.txt</code> file.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Specify the data folder that contains the classification output files from Matilda</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/classification/SHAREseq/test/"</span></span>
<span></span>
<span><span class="co"># Parse the accuracy_each_cell.txt file to get the real label and predicted label for comparison</span></span>
<span><span class="va">acc_each_cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="st">"accuracy_each_cell.txt"</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">acc_each_cell</span>, sep<span class="op">=</span><span class="st">'\t'</span><span class="op">)</span></span>
<span><span class="va">predicted_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="st">": "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>celltype<span class="op">=</span><span class="va">predicted_label</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">celltype</span>, fill<span class="op">=</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>Since we’ve previously classified our test set (i.e., the
<code>pbmc</code> dataset) using <code>scClassify</code> as well. We can
quickly have a look at the classification agreement between scClassify
and Matilda. In the heatmap below, the horizontal labels are cell types
for scClassify prediction while the vertical labels are the cell types
for Matilda prediction. Overall, we can see the clear agreement between
these two models along the diagonal. Besides, we can see that those
cells that are classified as unassigned by scClassify, are defined by
Matilda as different types of cells.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">predicted_label</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">annotation_scclassify</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">mat</span>, cluster_rows <span class="op">=</span> <span class="cn">F</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"Agreement between scClassify and Matilda"</span>, display_numbers <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="data-integration-and-dimension-reduction">Data integration and dimension reduction<a class="anchor" aria-label="anchor" href="#data-integration-and-dimension-reduction"></a>
</h4>
<p>Again, in our Python environment, we’ve run data integration and
dimension reduction on both evaluation set (with ground truth label) and
test set (without ground truth label). The output files are saved in
<code>data/output/dim_reduce/SHAREseq/eval/</code> for evaluation set
and <code>data/output/dim_reduce/SHAREseq/test/</code> for test set.
Both folder contain the <code>latent_space.csv</code> file, which stores
Matilda’s joint latent embedding for the data. Only for the evaluation
set, there is another output file named
<code>latent_space_label.csv</code>, which stores the cell type labels
for each cell in the data, which can be directly used for
visualization.</p>
<p>To inspect the integration results, we can use the following
<code>plot_latents</code> function, which will (1) conduct the UMAP
projection based on the Matilda’s joint latent embedding for
multimodality, and (2) conduct a simple k-means clustering and assess
cell type clustering on the Matilda’s joint latent embedding. More
specifically, 4 evaluation metrics are used for measuring the clustering
concordance between the clustering output and the cell type labels,
including Adjusted Rand Index (ARI), Normalized Mutual Information
(NMI), Fowlkes-Mallows index (FM), and Jaccard index (Jaccard). There
are two parameters for the <code>plot_latents</code> function:</p>
<ul>
<li><p><code>dimension_reduction_folder</code>: the folder which
contains the output files from running Matilda’s dimension reduction
task.</p></li>
<li><p><code>labels</code>: cell type labels for each cells in the data.
For evaluation set, we will ignore this parameter, in which case it will
automatically load the cell type labels from the
<code>latent_space_label.csv</code>. For the test set, we extract
prepare the list of cell types predicted by Matilda (as we did in the
cell type classification section above) and pass it to the
<code>labels</code> parameter.</p></li>
</ul>
<p>The returned value of <code>plot_latents</code> function contains (1)
a UMAP plot with cell type labeled for each cell, and (2) a performance
table which contains the results of the aforementioned 4 clustering
evaluation metrics.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot_latents</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dimension_reduction_folder</span>, <span class="va">labels</span><span class="op">=</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">latent_emb_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dimension_reduction_folder</span>, <span class="st">"latent_space.csv"</span><span class="op">)</span></span>
<span>  <span class="va">latent_emb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="va">latent_emb_csv</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">label_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dimension_reduction_folder</span>, <span class="st">"latent_space_label.csv"</span><span class="op">)</span></span>
<span>    <span class="va">latent_cty</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="va">label_csv</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span></span>
<span>    <span class="va">latent_cty</span> <span class="op">&lt;-</span> <span class="va">latent_cty</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">latent_cty</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">latent_cty</span> <span class="op">&lt;-</span> <span class="va">labels</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">latent_emb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">latent_emb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">seu_control</span><span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">latent_emb</span><span class="op">)</span></span>
<span>  <span class="va">seu_control</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">latent_cty</span></span>
<span>  <span class="va">seu_control</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">seu_control</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">seu_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu_control</span>,dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">seu_umap_dims</span> <span class="op">&lt;-</span> <span class="va">seu_control</span><span class="op">[[</span><span class="st">"umap"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">cell.embeddings</span></span>
<span>  <span class="va">kmeans_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">seu_umap_dims</span>, centers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">latent_cty</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/makeSNNGraph.html" class="external-link">makeSNNGraph</a></span><span class="op">(</span><span class="va">seu_umap_dims</span><span class="op">)</span></span>
<span>  <span class="va">communities</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_louvain.html" class="external-link">cluster_louvain</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span>
<span>  <span class="va">SNN_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">communities</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span>      </span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu_control</span>, reduction <span class="op">=</span> <span class="st">'umap'</span>,group.by <span class="op">=</span> <span class="st">'celltype'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">ari</span> <span class="op">&lt;-</span> <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">seu_control</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">kmeans_res</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> </span>
<span>  <span class="va">nmi</span> <span class="op">&lt;-</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/NMI.html" class="external-link">NMI</a></span><span class="op">(</span><span class="va">seu_control</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">kmeans_res</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> </span>
<span>  <span class="va">fm</span> <span class="op">&lt;-</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/AMI.html" class="external-link">AMI</a></span><span class="op">(</span><span class="va">seu_control</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">kmeans_res</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> </span>
<span>  <span class="va">jaccard</span> <span class="op">&lt;-</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/RI.html" class="external-link">RI</a></span><span class="op">(</span><span class="va">seu_control</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">kmeans_res</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> </span>
<span></span>
<span>  <span class="va">performance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ari<span class="op">=</span><span class="va">ari</span>, nmi<span class="op">=</span><span class="va">nmi</span>, fm<span class="op">=</span><span class="va">fm</span>, jac<span class="op">=</span><span class="va">jaccard</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">performance</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now, let’s first have a look at the evaluation set.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the data folder that contains the output files from running Matilda's data integration and dimension reduction task</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/dim_reduce/SHAREseq/eval/"</span></span>
<span></span>
<span><span class="co"># Plot the UMAP with cell types and calculate the clustering metrics</span></span>
<span><span class="va">latents</span> <span class="op">&lt;-</span> <span class="fu">plot_latents</span><span class="op">(</span>dimension_reduction_folder<span class="op">=</span><span class="va">data_folder</span><span class="op">)</span></span></code></pre></div>
<p>The first returned value is the UMAP projection with cell type
labeled. We can see that Matilda did good cell type separation under the
UMAP projection.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize the UMAP with cell types</span></span>
<span><span class="va">latents</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-55-1.png" width="700"></p>
<p>The second returned value is the performance table containing the 4
evaluation metrics for clustering.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize the performance table</span></span>
<span><span class="va">latents</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;         ari       nmi        fm       jac</span></span>
<span><span class="co">#&gt; 1 0.3919251 0.6153503 0.6126039 0.8906531</span></span></code></pre></div>
<p>Similarly, we can do the UMAP project for test set, together with the
cell types predicted by Matilda.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the data folder that contains the classification output files from Matilda</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/classification/SHAREseq/test/"</span></span>
<span></span>
<span><span class="co"># Parse the accuracy_each_cell.txt file to get the real label and predicted label for comparison</span></span>
<span><span class="va">acc_each_cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="st">"accuracy_each_cell.txt"</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">acc_each_cell</span>, sep<span class="op">=</span><span class="st">'\t'</span><span class="op">)</span></span>
<span><span class="va">predicted_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="st">": "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specify the data folder that contains the output files from running Matilda's data integration and dimension reduction task </span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="st">"../data/output/dim_reduce/SHAREseq/test/"</span></span>
<span></span>
<span><span class="co"># Plot the UMAP with cell types and calculate the clustering metrics</span></span>
<span><span class="va">latents</span> <span class="op">&lt;-</span> <span class="fu">plot_latents</span><span class="op">(</span>dimension_reduction_folder<span class="op">=</span><span class="va">data_folder</span>, labels<span class="op">=</span><span class="va">predicted_label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize the UMAP with cell types</span></span>
<span><span class="va">latents</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-58-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h4>
<p>Lastly, let’s take a look at the feature selection, which selects the
genes that are important for defining the specific cell type. The output
files from Matilda can be found from
<code>data/output/marker/SHAREseq/eval/</code>. We can see that, for
each cell type, there is a related csv file, which contains a list of
genes and its corresponding importance score calculated by Matilda. The
higher the score, the more important the gene.</p>
<p>Since we’ve run the feature selection on the evaluation set for
demonstration in our Python environment. Let’s first load the data
matrix and the cell type of our evaluation set, based on which, we then
run UMAP projection for visualizing the expression pattern of Matilda’s
selected genes.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the evaluation set data matrix for RNA</span></span>
<span><span class="va">h5f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="st">"../data/rna_pbmc_eval.h5"</span><span class="op">)</span></span>
<span><span class="va">data_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_rna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_rna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">barcodes</span></span>
<span></span>
<span><span class="co"># Load the evaluation set data matrix for ATAC</span></span>
<span><span class="va">h5f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="st">"../data/gas_pbmc_eval.h5"</span><span class="op">)</span></span>
<span><span class="va">data_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">h5f</span><span class="op">$</span><span class="va">matrix</span><span class="op">$</span><span class="va">barcodes</span></span>
<span></span>
<span></span>
<span><span class="co"># Load the cell types of evaluation set </span></span>
<span><span class="va">cty</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">'../data/pbmc_seurat_annotation_cty_eval.csv'</span><span class="op">)</span></span>
<span><span class="va">cty</span> <span class="op">&lt;-</span> <span class="va">cty</span><span class="op">$</span><span class="va">x</span></span>
<span></span>
<span><span class="co"># Run UMAP for RNA</span></span>
<span><span class="va">sce_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">data_rna</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_rna</span><span class="op">)</span></span>
<span><span class="va">sce_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_rna</span><span class="op">)</span></span>
<span><span class="va">umap_layout_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">sce_rna</span><span class="op">@</span><span class="va">int_colData</span><span class="op">$</span><span class="va">reducedDims</span><span class="op">$</span><span class="va">UMAP</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run UMAP for ATAC</span></span>
<span><span class="va">sce_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">data_atac</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_atac</span><span class="op">)</span></span>
<span><span class="va">sce_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_atac</span><span class="op">)</span></span>
<span><span class="va">umap_layout_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">sce_atac</span><span class="op">@</span><span class="va">int_colData</span><span class="op">$</span><span class="va">reducedDims</span><span class="op">$</span><span class="va">UMAP</span><span class="op">)</span></span></code></pre></div>
<p>In this example, let’s focus on CD14 Mono with the top 3
representative marker genes selected by Matilda. You can try other cell
types as well!</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the target cell type.</span></span>
<span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="st">'CD14 Mono'</span></span>
<span></span>
<span><span class="co"># Select the top 3 marker gene based on the importance score</span></span>
<span><span class="va">marker</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"../data/output/marker/SHAREseq/eval/fs.celltype_{cell_type}.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">top3</span><span class="op">&lt;-</span> <span class="va">marker</span><span class="op">$</span><span class="va">X</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">marker</span><span class="op">$</span><span class="va">importance.score</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Get the expression data of the top 3 marker genes </span></span>
<span><span class="va">top1_exp_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_rna</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="va">top2_exp_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_rna</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="va">top3_exp_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_rna</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">top1_exp_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_atac</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="va">top2_exp_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_atac</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="va">top3_exp_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce_atac</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">top3</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="op">]</span></span></code></pre></div>
<p>The following code visualizes the UMAP project with only the target
cell type labeled, which can be used for comparing the expression
pattern of the marker gene later.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the target cell type</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">cty</span><span class="op">==</span><span class="va">cell_type</span>, <span class="va">cty</span>,<span class="st">'others'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize the UMAP with target cell type only for RNA</span></span>
<span><span class="va">plot0_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_rna</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">cells</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"RNA - cell type: {cell_type} "</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize the UMAP with target cell type only for ATAC</span></span>
<span><span class="va">plot0_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_atac</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">cells</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"ATAC - cell type: {cell_type} "</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot0_rna</span><span class="op">+</span><span class="va">plot0_atac</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-61-1.png" width="700"></p>
<p>Now, we can plot the UMAP of marker gene expression, for the top 1
important gene:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># name of the marker gene </span></span>
<span><span class="va">marker_gene</span> <span class="op">&lt;-</span> <span class="va">top3</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top1_exp_rna</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for RNA</span></span>
<span><span class="va">plot_marker_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_rna</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"RNA - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top1_exp_atac</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for ATAC</span></span>
<span><span class="va">plot_marker_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_atac</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"ATAC - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">plot0_rna</span>, <span class="va">plot0_atac</span>, <span class="va">plot_marker_rna</span>, <span class="va">plot_marker_atac</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-62-1.png" width="700"></p>
<p>Similarly, we can plot the UMAP of marker gene expression, for the
second important gene:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># name of the marker gene </span></span>
<span><span class="va">marker_gene</span> <span class="op">&lt;-</span> <span class="va">top3</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top2_exp_rna</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for RNA</span></span>
<span><span class="va">plot_marker_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_rna</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"RNA - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top2_exp_atac</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for ATAC</span></span>
<span><span class="va">plot_marker_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_atac</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"ATAC - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">plot0_rna</span>, <span class="va">plot0_atac</span>, <span class="va">plot_marker_rna</span>, <span class="va">plot_marker_atac</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-63-1.png" width="700"></p>
<p>Similarly, we can plot the UMAP of marker gene expression, for the
third important gene:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># name of the marker gene </span></span>
<span><span class="va">marker_gene</span> <span class="op">&lt;-</span> <span class="va">top3</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top3_exp_rna</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for RNA</span></span>
<span><span class="va">plot_marker_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_rna</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"RNA - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># expression data of the marker gene </span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">top3_exp_atac</span></span>
<span></span>
<span><span class="co"># visualize the UMAP based on the marker gene expression for ATAC</span></span>
<span><span class="va">plot_marker_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_layout_atac</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>,<span class="va">UMAP2</span>,color<span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_colour_continuous</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">"red"</span>, low<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span>   <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span>,panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"ATAC - marker: {marker_gene}"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">plot0_rna</span>, <span class="va">plot0_atac</span>, <span class="va">plot_marker_rna</span>, <span class="va">plot_marker_atac</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_10x_multiome_files/figure-html/unnamed-chunk-64-1.png" width="700"></p>
<p>As expected, these visualizations reveal that features selected by
Matilda for each data modality show expression specificity towards their
respective cell types, demonstrating their potential usage for
characterizing cell identity and their underlying molecular
programs.</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-multiometech" class="csl-entry">
Baysoy, Alev, Zhiliang Bai, Rahul Satija, and Rong Fan. 2023. <span>“The
Technological Landscape and Applications of Single-Cell
Multi-Omics.”</span> <em>Nature Reviews Molecular Cell Biology</em> 24
(10): 695–713.
</div>
<div id="ref-encode" class="csl-entry">
Consortium, ENCODE Project et al. 2012. <span>“An Integrated
Encyclopedia of DNA Elements in the Human Genome.”</span>
<em>Nature</em> 489 (7414): 57.
</div>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carissa Chen, Sharon Long, Pengyi Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
